# 01. 초보자를 위한 파이썬 300제

---
## 001~010. print 시작하기

### 001. print 기초  
화면에 `Hello World` 문자열을 출력하세요.

In [1]:
print("Hello World");

Hello World


### 002. print 기초  
화면에 `Mary's cosmetics`을 출력하세요. (중간에 '가 있음에 주의하세요)

In [2]:
print("Mary's cosmetics")

Mary's cosmetics


### 003. print 기초  
화면에 아래 문장을 출력하세요. (중간에 "가 있음에 주의하세요.)
> 신씨가 소리질렀다. "도둑이야".

In [3]:
print("신씨가 소리질렀다. \"도둑이야\".")
print('신씨가 소리질렀다. "도둑이야".')

신씨가 소리질렀다. "도둑이야".
신씨가 소리질렀다. "도둑이야".


### 004. print 기초  
화면에 `C:\Windows`를 출력하세요.

In [4]:
print(r"C:\Windows")
print("C:\\Windows")

C:\Windows
C:\Windows


### 005. print 탭과 줄바꿈  
다음 코드를 실행해보고 `\t`와 `\n`의 역할을 설명해보세요.
```python
print("안녕하세요.\n만나서\t\t반갑습니다.")
```

In [5]:
# 줄바꿈(\n), 탭(\t)
print("안녕하세요.\n만나서\t\t반갑습니다.")

안녕하세요.
만나서		반갑습니다.


### 006. print 여러 데이터 출력  
print 함수에 두 개의 단어를 입력한 예제입니다. 아래 코드의 출력 결과를 예상해봅시다.
```python
print ("오늘은", "일요일")
```

In [6]:
# 오늘은 일요일
print ("오늘은", "일요일")

오늘은 일요일


### 007. print 기초  
print() 함수를 사용하여 다음과 같이 출력하세요. (sep 이용)
> naver;kakao;sk;samsung

In [7]:
print("naver", "kakao", "sk", "samsung", sep=";")

naver;kakao;sk;samsung


### 008. print 기초  
print() 함수를 사용하여 다음과 같이 출력하세요.
> naver/kakao/sk/samsung

In [8]:
print("naver", "kakao", "sk", "samsung", sep="\\")

naver\kakao\sk\samsung


### 009. print 줄바꿈  
다음 코드를 수정하여 줄바꿈이 없이 출력하세요.  
(힌트: end='') print 함수는 두 번 사용합니다. 세미콜론 (;)은 한줄에 여러 개의 명령을 작성하기 위해 사용합니다. 
```python
print("first"); print("second")
```

In [9]:
print("first", end=" "); print("second")

first second


### 010. 연산 결과 출력  
5/3의 결과를 화면에 출력하세요.

In [10]:
print("5/3 =", 5/3)

5/3 = 1.6666666666666667


---
## 011~020. 변수  

### 011. 변수 사용하기  
삼성전자라는 변수로 50,000원을 바인딩해보세요. 삼성전자 주식 10주를 보유하고 있을 때 총 평가금액을 출력하세요.

In [11]:
삼성전자 = 50000
print("평가금액:",format(삼성전자*10, ','), "원")
print(f"평가금액: {삼성전자*10 :,} 원")

평가금액: 500,000 원
평가금액: 500,000 원


### 012. 변수 사용하기  
다음 표는 삼성전자의 일부 투자정보입니다. 변수를 사용해서 시가총액, 현재가, PER 등을 바인딩해보세요.
|항목|값|
|:--|--:|
|시가총액|298조|
|현재가|50,000원|
|PER|15.79|

In [12]:
시가총액 = 298 * 10**12
현재가 = 50000
PER = 15.79
print(f"시가총액: {시가총액:,} 원 \n현재가: {현재가:,} 원 \nPER: {PER}")

시가총액: 298,000,000,000,000 원 
현재가: 50,000 원 
PER: 15.79


### 013. 문자열 출력
변수 s와 t에는 각각 문자열이 바인딩 되어있습니다.
```python
s = "hello"  
t = "python"  
```
두 변수를 이용하여 아래와 같이 출력해보세요.  
실행 예:  
> hello! python  

In [13]:
s = "hello"
t = "python"
print(s+"!",t)
print(f"{s}! {t}")

hello! python
hello! python


### 014. 파이썬을 이용한 값 계산
아래 코드의 실행 결과를 예상해보세요.  
```python
2 + 2 * 3 
```

In [14]:
# 8
print(2 + 2 * 3)

8


### 015. type 함수  
type() 함수는 데이터 타입을 판별합니다. 변수 a에는 128 숫자가 바인딩돼 있어 type 함수가 int (정수)형임을 알려줍니다.  
```python
a = 128  
print (type(a))  
```
> <class 'int'>

아래 변수에 바인딩된 값의 타입을 판별해보세요.  
```python
a = "132"
```

In [15]:
a = "132"
print(type(a))

<class 'str'>


### 016. 문자열을 정수로 변환  
문자열 '720'를 정수형으로 변환해보세요.  
```python
num_str = "720"
```

In [16]:
num_str = "720"
num = int(num_str)
print(type(num))

<class 'int'>


### 017. 정수를 문자열 100으로 변환
정수 100을 문자열 '100'으로 변환해보세요.
```python
num = 100
```

In [17]:
num = 100
num_str = str(num)
print(type(num_str))

<class 'str'>


### 018. 문자열을 실수로 변환  
문자열 "15.79"를 실수(float) 타입으로 변환해보세요.

In [18]:
num_str = "15.79"
num = float(num_str)
print(type(num))

<class 'float'>


### 019. 문자열을 정수로 변환  
year라는 변수가 문자열 타입의 연도를 바인딩하고 있습니다. 이를 정수로 변환한 후 최근 3년의 연도를 화면에 출력해보세요.  
```python
year = "2020"
```

In [19]:
year = "2020"
num_year = int(year)
print(num_year-3, num_year-2, num_year-1)

2017 2018 2019


### 020. 파이썬 계산  
에어컨이 월 48,584원에 무이자 36개월의 조건으로 홈쇼핑에서 판매되고 있습니다. 총 금액은 계산한 후 이를 화면에 출력해보세요. (변수사용하기)  

In [20]:
price_per_month = 48584
months = 36
print(f"총 금액: {price_per_month * months:,} 원")

총 금액: 1,749,024 원


---
## 021~050. 파이썬 문자열  

### 021. 문자열 인덱싱  
letters가 바인딩하는 문자열에서 첫번째와 세번째 문자를 출력하세요.
```python
letters = 'python'
```
실행 예:  
> p t  

In [21]:
letters = 'python'
print(letters[0], letters[2])

p t


### 022. 문자열 슬라이싱  
자동차 번호가 다음과 같을 때 뒤에 4자리만 출력하세요.  
```python
license_plate = "24가 2210"
```
실행 예: 
> 2210  

In [22]:
license_plate = "24가 2210"
print(license_plate[-4:])

2210


### 023. 문자열 인덱싱  
아래의 문자열에서 '홀' 만 출력하세요.  
```python
string = "홀짝홀짝홀짝"  
```
실행 예:  
> 홀홀홀

In [23]:
string = "홀짝홀짝홀짝" 
print(string[::2])

홀홀홀


### 024. 문자열 슬라이싱  
문자열을 거꾸로 뒤집어 출력하세요.  
```python
string = "PYTHON"
```
실행 예:  
> NOHTYP  

In [24]:
string = "PYTHON"
print(string[::-1])

NOHTYP


### 025. 문자열 치환  
아래의 전화번호에서 하이푼 ('-')을 제거하고 출력하세요.  
```python
phone_number = "010-1111-2222"  
```
실행 예:  
> 010 1111 2222  

In [25]:
phone_number = "010-1111-2222"
print(phone_number.replace('-', ' '))

010 1111 2222


### 026. 문자열 다루기  
25번 문제의 전화번호를 아래와 같이 모두 붙여 출력하세요.  
실행 예:  
> 01011112222  

In [26]:
phone_number = "010-1111-2222"
print(phone_number.replace('-', ''))

01011112222


### 027. 문자열 다루기  
url 에 저장된 웹 페이지 주소에서 도메인을 출력하세요.  
```python
url = "http://sharebook.kr"
```
실행 예: 
> kr

In [27]:
url = "http://sharebook.kr"
print(url.split('.')[-1])

kr


### 028. 문자열은 immutable  
아래 코드의 실행 결과를 예상해보세요.  
```python
lang = 'python'  
lang[0] = 'P'  
print(lang)  
```

In [28]:
# TypeError
lang = 'python'
lang[0] = 'P'
print(lang)

TypeError: 'str' object does not support item assignment

### 029. replace 메서드  
아래 문자열에서 소문자 'a'를 대문자 'A'로 변경하세요.  
```python
string = 'abcdfe2a354a32a'
```
실행 예:  
> Abcdfe2A354A32A   

In [ ]:
string = 'abcdfe2a354a32a'
print(string.replace('a', 'A'))

### 030. replace 메서드  
아래 코드의 실행 결과를 예상해보세요.  
```python
string = 'abcd'  
string.replace('b', 'B')  
print(string)
```

In [ ]:
# abcd (원본변경불가)
string = 'abcd'
string.replace('b', 'B')
print(string)

### 031. 문자열 합치기  
아래 코드의 실행 결과를 예상해보세요.  
```python
a = "3"  
b = "4"  
print(a + b)  
```

In [ ]:
# "34"
a = "3"
b = "4"
print(a + b)

### 032. 문자열 곱하기  
아래 코드의 실행 결과를 예상해보세요.
```python
print("Hi" * 3)  
```

In [ ]:
# HiHiHi
print("Hi" * 3)

### 033. 문자열 곱하기  
화면에 `'-'를 80개` 출력하세요.  
실행 예:  
> <pre>--------------------------------------------------------------------------------</pre>

In [ ]:
print('-' * 80)

### 034. 문자열 곱하기  
변수에 다음과 같은 문자열이 바인딩되어 있습니다.  
```python
t1 = 'python'  
t2 = 'java'  
```
변수에 문자열 더하기와 문자열 곱하기를 사용해서 아래와 같이 출력해보세요.  
실행 예:  
> python java python java python java python java  

In [ ]:
t1 = 'python'
t2 = 'java'
print((t1 + ' ' + t2 + ' ') * 4)

### 035. 문자열 출력  
변수에 다음과 같이 문자열과 정수가 바인딩되어 있을 때 `% formatting`을 사용해서 다음과 같이 출력해보세요.  
```python
name1 = "김민수"  
age1 = 10  
name2 = "이철희"  
age2 = 13
```
실행 예:
> 이름: 김민수 나이: 10  
> 이름: 이철희 나이: 13  

In [ ]:
name1 = "김민수" 
age1 = 10
name2 = "이철희"
age2 = 13
print('이름: %s 나이: %d' % (name1, age1))
print('이름: %s 나이: %d' % (name2, age2))

### 036. 문자열 출력  
문자열의 `format( )`메서드를 사용해서 035번 문제를 다시 풀어보세요.  

In [ ]:
print('이름: {} 나이: {}'.format(name1, age1))
print('이름: {} 나이: {}'.format(name2, age2))

### 037. 문자열 출력  
파이썬 3.6부터 지원하는 `f-string`을 사용해서 035번 문제를 다시 풀어보세요.  

In [ ]:
print(f'이름: {name1} 나이: {age1}')
print(f'이름: {name2} 나이: {age2}')

### 038. 컴마 제거하기  
삼성전자의 상장주식수가 다음과 같습니다. 컴마를 제거한 후 이를 정수 타입으로 변환해보세요.  
```python
상장주식수 = "5,969,782,550"  
```

In [ ]:
상장주식수 = "5,969,782,550"
num = int(상장주식수.replace(',', ''))
print(num, type(num))

### 039. 문자열 슬라이싱  
다음과 같은 문자열에서 `'2020/03'`만 출력하세요.  
```python
분기 = "2020/03(E) (IFRS연결)"  
```

In [ ]:
분기 = "2020/03(E) (IFRS연결)"
print(분기[:7])
print(분기[:분기.index('(')])

### 040. strip 메서드  
문자열의 좌우의 공백이 있을 때 이를 제거해보세요.  
```python
data = "   삼성전자    "
```

In [ ]:
data = "   삼성전자    "
print(data.strip())

### 041. upper 메서드  
다음과 같은 문자열이 있을 때 이를 대문자 BTC_KRW로 변경하세요.  
```python
ticker = "btc_krw"  
```

In [ ]:
ticker = "btc_krw"
print(ticker.upper())

### 042. lower 메서드  
다음과 같은 문자열이 있을 때 이를 소문자 btc_krw로 변경하세요. 
```python
ticker = "BTC_KRW"  
```

In [ ]:
ticker = "BTC_KRW"
print(ticker.lower())

### 043. capitalize 메서드  
문자열 'hello'가 있을 때 이를 'Hello'로 변경해보세요.  

In [ ]:
print('hello'.capitalize())

### 044. endswith 메서드  
파일 이름이 문자열로 저장되어 있을 때 endswith 메서드를 사용해서 파일 이름이 'xlsx'로 끝나는지 확인해보세요.  
```python
file_name = "보고서.xlsx"  
```

In [ ]:
file_name = "보고서.xlsx"
print(file_name.endswith('xlsx'))

### 045. endswith 메서드  
파일 이름이 문자열로 저장되어 있을 때 endswith 메서드를 사용해서 파일 이름이 'xlsx' 또는 'xls'로 끝나는지 확인해보세요.  
```python
file_name = "보고서.xlsx"  
```

In [ ]:
file_name = "보고서.xlsx"
print(file_name.endswith('xlsx') or file_name.endswith('xls'))
print(file_name.endswith(('xls', 'xlsx')))

### 046. startswith 메서드  
파일 이름이 문자열로 저장되어 있을 때 startswith 메서드를 사용해서 파일 이름이 '2020'로 시작하는지 확인해보세요.  
```python
file_name = "2020_보고서.xlsx"  
```

In [ ]:
file_name = "2020_보고서.xlsx"
print(file_name.startswith('2020'))

### 047. split 메서드  
다음과 같은 문자열이 있을 때 공백을 기준으로 문자열을 나눠보세요.  
```python
a = "hello world"  
```

In [ ]:
a = "hello world"
print(a.split(' '))

### 048. split 메서드  
다음과 같이 문자열이 있을 때 btc와 krw로 나눠보세요.  
```python
ticker = "btc_krw"  
```

In [ ]:
ticker = "btc_krw"
print(ticker.split('_'))

### 049. split 메서드  
다음과 같이 날짜를 표현하는 문자열이 있을 때 연도, 월, 일로 나눠보세요.  
```python
date = "2020-05-01"  
```

In [ ]:
date = "2020-05-01"
print(date.split('-'))

### 050. rstrip 메서드  
문자열의 오른쪽에 공백이 있을 때 이를 제거해보세요.  
```python
data = "039490     "
```

In [ ]:
data = "039490     "
print(data.rstrip())

---
## 051~070. 파이썬 리스트  

### 051. 리스트 생성  
2016년 11월 영화 예매 순위 기준 top3는 다음과 같습니다. 영화 제목을 movie_rank 이름의 리스트에 저장해보세요. (순위 정보는 저장하지 않습니다.)  

|순위|영화|
|--|--|
|1|닥터 스트레인지|
|2|스플릿|
|3|럭키|

In [37]:
movie_rank = ['닥터 스트레인지', '스플릿', '럭키']
print(movie_rank)

['닥터 스트레인지', '스플릿', '럭키']


### 052. 리스트 원소 추가
051의 movie_rank 리스트에 "배트맨"을 추가하라.

In [38]:
movie_rank = ['닥터 스트레인지', '스플릿', '럭키']
movie_rank.append('배트맨')
print(movie_rank)

['닥터 스트레인지', '스플릿', '럭키', '배트맨']


### 053. 리스트 원소 추가 
movie_rank 리스트에는 아래와 같이 네 개의 영화 제목이 바인딩되어 있다. "슈퍼맨"을 "닥터 스트레인지"와 "스플릿" 사이에 추가하라.  
```python
movie_rank = ['닥터 스트레인지', '스플릿', '럭키', '배트맨']  
```

In [39]:
movie_rank = ['닥터 스트레인지', '스플릿', '럭키', '배트맨']
movie_rank.insert(1, '슈퍼맨')
print(movie_rank)

['닥터 스트레인지', '슈퍼맨', '스플릿', '럭키', '배트맨']


### 054. 리스트 원소 삭제
movie_rank 리스트에서 '럭키'를 삭제하라.  
```python
movie_rank = ['닥터 스트레인지', '슈퍼맨', '스플릿', '럭키', '배트맨']
```

In [40]:
movie_rank = ['닥터 스트레인지', '슈퍼맨', '스플릿', '럭키', '배트맨']
movie_rank.remove('럭키')
print(movie_rank)

movie_rank = ['닥터 스트레인지', '슈퍼맨', '스플릿', '럭키', '배트맨']
del movie_rank[3]
print(movie_rank)

['닥터 스트레인지', '슈퍼맨', '스플릿', '배트맨']
['닥터 스트레인지', '슈퍼맨', '스플릿', '배트맨']


### 055. 리스트 원소 삭제
movie_rank 리스트에서 '스플릿' 과 '배트맨'을 삭제하라.  
```python
movie_rank = ['닥터 스트레인지', '슈퍼맨', '스플릿', '배트맨']  
```

In [41]:
movie_rank = ['닥터 스트레인지', '슈퍼맨', '스플릿', '배트맨']
del movie_rank[2:]
print(movie_rank)

['닥터 스트레인지', '슈퍼맨']


### 056. 리스트 원소 결합  
lang1과 lang2 리스트가 있을 때 lang1과 lang2의 원소를 모두 갖고 있는 langs 리스트를 만들어라.  
```python
lang1 = ["C", "C++", "JAVA"]  
lang2 = ["Python", "Go", "C#"]
```
실행 예:  
```python
langs = ['C', 'C++', 'JAVA', 'Python', 'Go', 'C#']  
```

In [44]:
lang1 = ["C", "C++", "JAVA"]
lang2 = ["Python", "Go", "C#"]
langs = lang1 + lang2
print(langs)

['C', 'C++', 'JAVA', 'Python', 'Go', 'C#']


### 057. 리스트 원소 최대, 최소  
다음 리스트에서 최댓값과 최솟값을 출력하라. (힌트: min(), max() 함수 사용)  
```python
nums = [1, 2, 3, 4, 5, 6, 7]  
```
실행 예:  
> max:  7  
> min:  1  

In [47]:
nums = [1, 2, 3, 4, 5, 6, 7]
print('max:', max(nums))
print('min:', min(nums))

max: 7
min: 1


### 058. 리스트 원소 합  
다음 리스트의 합을 출력하라.  
```python
nums = [1, 2, 3, 4, 5]  
```
실행 예:
> 15

In [48]:
nums = [1, 2, 3, 4, 5]
print(sum(nums))

15


### 059. 리스트 원소 개수  
다음 리스트에 저장된 데이터의 개수를 화면에 구하하라.  
```python
cook = ["피자", "김밥", "만두", "양념치킨", "족발", "피자", "김치만두", "쫄면", "소시지", "라면", "팥빙수", "김치전"]  
```

In [49]:
cook = ["피자", "김밥", "만두", "양념치킨", "족발", "피자", "김치만두", "쫄면", "소시지", "라면", "팥빙수", "김치전"]
print(len(cook))

12


### 060. 리스트 원소 평균
다음 리스트의 평균을 출력하라.  
```python
nums = [1, 2, 3, 4, 5]
```
실행 예:  
> 3.0

In [51]:
nums = [1, 2, 3, 4, 5]
print(sum(nums)/len(nums))

3.0


### 061. 리스트 슬라이싱  
price 변수에는 날짜와 종가 정보가 저장돼 있다. 날짜 정보를 제외하고 가격 정보만을 출력하라. (힌트 : 슬라이싱)  
```python
price = ['20180728', 100, 130, 140, 150, 160, 170]  
```
출력 예:
> [100, 130, 140, 150, 160, 170]

In [52]:
price = ['20180728', 100, 130, 140, 150, 160, 170]
print(price[1:])

[100, 130, 140, 150, 160, 170]


### 062. 리스트 슬라이싱  
슬라이싱을 사용해서 홀수만 출력하라.  
```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
실행 예:  
>  [1, 3, 5, 7, 9]  

In [53]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(nums[::2])

[1, 3, 5, 7, 9]


### 063. 리스트 슬라이싱  
슬라이싱을 사용해서 짝수만 출력하라.  
```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  
```
실행 예:
>  [2, 4, 6, 8, 10]

In [54]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(nums[1::2])

[2, 4, 6, 8, 10]


### 064. 리스트 슬라이싱
슬라이싱을 사용해서 리스트의 숫자를 역 방향으로 출력하라.  
```python
nums = [1, 2, 3, 4, 5]
```
실행 예:  
> [5, 4, 3, 2, 1]  

In [62]:
nums = [1, 2, 3, 4, 5]
print(nums[::-1])
print(list(reversed(nums)))

[5, 4, 3, 2, 1]
[5, 4, 3, 2, 1]


### 065. 리스트 원소 추출
interest 리스트에는 아래의 데이터가 바인딩되어 있다.
```python
interest = ['삼성전자', 'LG전자', 'Naver']
```
interest 리스트를 사용하여 아래와 같이 화면에 출력하라.  
출력 예:  
> 삼성전자 Naver  

In [63]:
interest = ['삼성전자', 'LG전자', 'Naver']
print(interest[0], interest[2])

삼성전자 Naver


### 066. join 메서드
interest 리스트에는 아래의 데이터가 바인딩되어 있다.  
```python
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']  
```
interest 리스트를 사용하여 아래와 같이 화면에 출력하라.  
출력 예:  
> 삼성전자 LG전자 Naver SK하이닉스 미래에셋대우  

In [65]:
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']
print(' '.join(interest))

삼성전자 LG전자 Naver SK하이닉스 미래에셋대우


### 067 join 메서드  
interest 리스트에는 아래의 데이터가 바인딩되어 있다.  
```python
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']
```
interest 리스트를 사용하여 아래와 같이 화면에 출력하라.  
출력 예:  
> 삼성전자/LG전자/Naver/SK하이닉스/미래에셋대우  

In [66]:
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']
print('/'.join(interest))

삼성전자/LG전자/Naver/SK하이닉스/미래에셋대우


### 068. join 메서드  
interest 리스트에는 아래의 데이터가 바인딩되어 있다.  
```python
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']  
```
join() 메서드를 사용해서 interest 리스트를 아래와 같이 화면에 출력하라.  
출력 예:
> 삼성전자  
> LG전자  
> Naver  
> SK하이닉스  
> 미래에셋대우  

In [67]:
interest = ['삼성전자', 'LG전자', 'Naver', 'SK하이닉스', '미래에셋대우']
print('\n'.join(interest))

삼성전자
LG전자
Naver
SK하이닉스
미래에셋대우


### 069. 문자열 split 메서드  
회사 이름이 슬래시 ('/')로 구분되어 하나의 문자열로 저장되어 있다.  
```python
string = "삼성전자/LG전자/Naver"  
```
이를 interest 이름의 리스트로 분리 저장하라.  
실행 예:  
```python
print(interest)  
```
> ['삼성전자', 'LG전자', 'Naver']  

In [68]:
string = "삼성전자/LG전자/Naver"
interest = string.split('/')
print(interest)

['삼성전자', 'LG전자', 'Naver']


### 070. 리스트 정렬  
리스트에 있는 값을 오름차순으로 정렬하세요.  
```python
data = [2, 4, 3, 1, 5, 10, 9]  
```

In [71]:
data = [2, 4, 3, 1, 5, 10, 9]
print(sorted(data))
data.sort()
print(data)

[1, 2, 3, 4, 5, 9, 10]
[1, 2, 3, 4, 5, 9, 10]


---
## 071~080. 파이썬 튜플

### 071. 튜플 생성
my_variable 이름의 비어있는 튜플을 만들라.  

In [74]:
my_variable = ()
print(my_variable, type(my_variable))

() <class 'tuple'>


### 072. 튜플 생성
2016년 11월 영화 예매 순위 기준 top3는 다음과 같다. 영화 제목을 movie_rank 이름의 튜플에 저장하라. (순위 정보는 저장하지 않는다.)  

|순위|영화|
|--|--|
|1|닥터 스트레인지|
|2|스플릿|
|3|럭키|

In [75]:
movie_rank = ('닥터 스트레인지', '스플릿', '럭키')
print(movie_rank)

('닥터 스트레인지', '스플릿', '럭키')


### 073. 튜플 생성
숫자 1 이 저장된 튜플을 생성하라.

In [76]:
my_variable = (1,)
print(my_variable, type(my_variable))

(1,) <class 'tuple'>


### 074. 튜플 불변
다음 코드를 실행해보고 오류가 발생하는 원인을 설명하라.  
```python
t = (1, 2, 3)  
t[0] = 'a'  
```
> Traceback (most recent call last):  
>   File "<pyshell#46>", line 1, in <module>  
>     t[0] = 'a'  
> TypeError: 'tuple' object does not support item assignment  

In [77]:
# 튜플은 불변형, 수정불가
t = (1,2,3)
t[0] = 'a'

TypeError: 'tuple' object does not support item assignment

### 075. 튜플 생성
아래와 같이 t에는 1, 2, 3, 4 데이터가 바인딩되어 있다. t가 바인딩하는 데이터 타입은 무엇인가?  
```python
t = 1, 2, 3, 4  
```

In [79]:
# 튜플, 소괄호 () 없이 선언 가능
t = 1, 2, 3, 4
print(type(t))

<class 'tuple'>


### 076. 튜플 생성
변수 t에는 아래와 같은 값이 저장되어 있다. 변수 t가 ('A', 'b', 'c') 튜플을 가리키도록 수정 하라.  
```python
t = ('a', 'b', 'c')
```

In [82]:
t = ('a', 'b', 'c')
t = t[0].upper(), *t[1:]
print(t)

('A', 'b', 'c')


### 077. 튜플 변환
다음 튜플을 리스트로 변환하라.  
```python
interest = ('삼성전자', 'LG전자', 'SK Hynix')  
```

In [85]:
interest = ('삼성전자', 'LG전자', 'SK Hynix')
interest_list = list(interest)
print(interest_list, type(interest_list))

['삼성전자', 'LG전자', 'SK Hynix'] <class 'list'>


### 078. 튜플 변환
다음 리스트를 튜플로 변경하라.  
```python
interest = ['삼성전자', 'LG전자', 'SK Hynix']
```

In [86]:
interest = ['삼성전자', 'LG전자', 'SK Hynix']
interest_tuple = tuple(interest)
print(interest_tuple, type(interest_tuple))

('삼성전자', 'LG전자', 'SK Hynix') <class 'tuple'>


### 079. 튜플 언팩킹
다음 코드의 실행 결과를 예상하라.  
```python
temp = ('apple', 'banana', 'cake')  
a, b, c = temp  
print(a, b, c)  
```

In [87]:
# apple banana cake
temp = ('apple', 'banana', 'cake')
a, b, c = temp
print(a, b, c)

apple banana cake


### 080. range 함수  
1 부터 99까지의 정수 중 짝수만 저장된 튜플을 생성하라.  
> (2, 4, 6, 8 ... 98)  

In [90]:
print(tuple(range(2,99,2)))

(2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98)


---
## 081~100. 파이썬 딕셔너리

### 081. 별 표현식
기본적으로 데이터 언패킹은 좌변의 변수와 우변 데이터 개수가 같아야 합니다. 하지만 star expression을 사용하면 변수의 개수가 달라도 데이터 언패킹을 할 수 있습니다. 튜플에 저장된 데이터 중에서 앞에 있는 두 개의 데이터만 필요할 경우 나머지 데이터의 언패킹 코드를 작성할 필요가 없습니다.  
```python
a, b, *c = (0, 1, 2, 3, 4, 5)  
a  
```
> 0  
```python
b  
```
> 1  
```python
c  
```
> [2, 3, 4, 5]  

다음과 같이 10개의 값이 저장된 scores 리스트가 있을 때, star expression을 사용하여 좌측 8개의 값을 valid_score 변수에 바인딩하여라.  
```python
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]  
```

In [91]:
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]
*valid_score, _, _ = scores;
print(valid_score)

[8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5]


### 082. 별 표현식 
다음과 같이 10개의 값이 저장된 scores 리스트가 있을 때, star expression을 사용하여 우측 8개의 값을 valid_score 변수에 바인딩하여라. 
```python
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]  
```

In [92]:
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]
_, _, *valid_score = scores
print(valid_score)

[8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]


### 083. 별 표현식  
다음과 같이 10개의 값이 저장된 scores 리스트가 있을 때, star expression을 사용하여 가운데 있는 8개의 값을 valid_score 변수에 바인딩하여라.  
```python
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]  
```

In [93]:
scores = [8.8, 8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8, 9.4]
_, *valid_score, _ = scores
print(valid_score)

[8.9, 8.7, 9.2, 9.3, 9.7, 9.9, 9.5, 7.8]


### 084. 비어있는 딕셔너리
temp 이름의 비어있는 딕셔너리를 만들라.

In [94]:
temp = {}
print(temp, type(temp))

{} <class 'dict'>


### 085. 딕셔너리 생성
다음 아이스크림 이름과 희망 가격을 딕셔너리로 구성하라.  

|이름|희망 가격|
|--|--|
|메로나|1000|
|폴라포|1200|
|빵빠레|1800|

In [98]:
icecream = { '이름': ['메로나','폴라포','빵빠레'], '희망 가격': [1000,1200,1800]}
print(icecream)
icecream = { '메로나':1000, '폴라포':1200, '빵빠레':1800 }
print(icecream)

{'이름': ['메로나', '폴라포', '빵빠레'], '희망 가격': [1000, 1200, 1800]}
{'메로나': 1000, '폴라포': 1200, '빵빠레': 1800}


### 086. 딕셔너리 추가
085 번의 딕셔너리에 아래 아이스크림 가격정보를 추가하라.  

|이름|희망 가격|
|--|--|
|죠스바|1200|
|월드콘|1500|

In [99]:
icecream = { '메로나':1000, '폴라포':1200, '빵빠레':1800 }
icecream['죠스바'] = 1200
icecream['월드콘'] = 1500
print(icecream)

{'메로나': 1000, '폴라포': 1200, '빵빠레': 1800, '죠스바': 1200, '월드콘': 1500}


### 087. 딕셔너리 추출
다음 딕셔너리를 사용하여 메로나 가격을 출력하라.  

```python
ice = {'메로나': 1000,  
       '폴로포': 1200,  
       '빵빠레': 1800,  
       '죠스바': 1200,  
       '월드콘': 1500}  
```

실행 예:  
> 메로나 가격: 1000  

In [100]:
ice = {'메로나': 1000,
       '폴로포': 1200,
       '빵빠레': 1800,
       '죠스바': 1200,
       '월드콘': 1500}
print(f"메로나 가격: {ice['메로나']}")

메로나 가격: 1000


### 088. 딕셔너리 수정
다음 딕셔너리에서 메로나의 가격을 1300으로 수정하라.  
```python
ice = {'메로나': 1000,  
       '폴로포': 1200,  
       '빵빠레': 1800,  
       '죠스바': 1200,  
       '월드콘': 1500}  
```

In [102]:
ice = {'메로나': 1000,
       '폴로포': 1200,
       '빵빠레': 1800,
       '죠스바': 1200,
       '월드콘': 1500}
ice['메로나'] = 1300
print(ice)

{'메로나': 1300, '폴로포': 1200, '빵빠레': 1800, '죠스바': 1200, '월드콘': 1500}


### 089. 딕셔너리 삭제  
다음 딕셔너리에서 메로나를 삭제하라.  
```python
ice = {'메로나': 1000,  
       '폴로포': 1200,  
       '빵빠레': 1800,  
       '죠스바': 1200,  
       '월드콘': 1500}  
```

In [103]:
ice = {'메로나': 1000,
       '폴로포': 1200,
       '빵빠레': 1800,
       '죠스바': 1200,
       '월드콘': 1500}
del ice['메로나']
print(ice)

{'폴로포': 1200, '빵빠레': 1800, '죠스바': 1200, '월드콘': 1500}


### 090. 딕셔너리 키값 존재여부
다음 코드에서 에러가 발생한 원인을 설명하라.  
```python
icecream = {'폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}  
icecream['누가바']  
```
출력 예:
> Traceback (most recent call last):  
>   File "<pyshell#69>", line 1, in <module>  
>     icecream['누가바']  
> KeyError: '누가바'  

In [104]:
# 딕셔너리의 키값이 존재하지 않음
icecream = {'폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}
icecream['누가바']

KeyError: '누가바'

### 091. 딕셔너리 생성  
아래의 표에서, 아이스크림 이름을 키값으로, (가격, 재고) 리스트를 딕셔너리의 값으로 저장하라. 딕셔너리의 이름은 inventory로 한다.  

|이름|가격|재고|
|--|--|--|
|메로나|300|20|
|비비빅|400|3|
|죠스바|250|100|

In [105]:
inventory = {
    '메로나': [300, 20],
    '비비빅': [400, 3],
    '죠스바': [250, 100],
}
print(inventory)

{'메로나': [300, 20], '비비빅': [400, 3], '죠스바': [250, 100]}


### 092. 딕셔너리 인덱싱
inventory 딕셔너리에서 메로나의 가격을 화면에 출력하라.  
```python
inventory = {"메로나": [300, 20],  
              "비비빅": [400, 3],  
              "죠스바": [250, 100]}  
```
실행 예:  
> 300 원  

In [107]:
inventory = {
    '메로나': [300, 20],
    '비비빅': [400, 3],
    '죠스바': [250, 100],
}
print(inventory['메로나'][0], '원')

300 원


### 093. 딕셔너리 인덱싱  
inventory 딕셔너리에서 메로나의 재고를 화면에 출력하라.  
```python
inventory = {"메로나": [300, 20],  
              "비비빅": [400, 3],  
              "죠스바": [250, 100]}  
```
실행 예:  
> 20 개  

In [108]:
inventory = {
    '메로나': [300, 20],
    '비비빅': [400, 3],
    '죠스바': [250, 100],
}
print(inventory['메로나'][1], '개')

20 개


### 094. 딕셔너리 추가  
```python
inventory = {"메로나": [300, 20],  
              "비비빅": [400, 3],  
              "죠스바": [250, 100]}  
```
|이름|가격|재고|
|--|--|--|
|월드콘|500|7|
실행 예:  
```python
print(inventory)  
```
> {'메로나': [300, 20], '비비빅': [400, 3], '죠스바': [250, 100], '월드콘': [500, 7]}

In [109]:
inventory = {
    '메로나': [300, 20],
    '비비빅': [400, 3],
    '죠스바': [250, 100],
}
inventory['월드콘'] = [500, 7]
print(inventory)

{'메로나': [300, 20], '비비빅': [400, 3], '죠스바': [250, 100], '월드콘': [500, 7]}


### 095. 딕셔너리 keys() 메서드  
다음의 딕셔너리로부터 key 값으로만 구성된 리스트를 생성하라.  
```python
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}  
```

In [110]:
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}
print(icecream.keys())

dict_keys(['탱크보이', '폴라포', '빵빠레', '월드콘', '메로나'])


### 096. 딕셔너리 values() 메서드  
다음의 딕셔너리에서 values 값으로만 구성된 리스트를 생성하라.  
```python
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}  
```

In [111]:
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}
print(icecream.values())

dict_values([1200, 1200, 1800, 1500, 1000])


### 097. 딕셔너리 values() 메서드
icecream 딕셔너리에서 아이스크림 판매 금액의 총합을 출력하라.  
```python
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}  
```
출력 예:  
> 6700

In [112]:
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}
print(sum(icecream.values()))

6700


### 098. 딕셔너리 update 메서드  
아래의 new_product 딕셔너리를 다음 icecream 딕셔너리에 추가하라. 
```python
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}  
new_product = {'팥빙수':2700, '아맛나':1000}  
```
실행 예:  
```python
print(icecream)  
```
> {'탱크보이': 1200,  '폴라포': 1200,  '빵빠레': 1800,  '월드콘': 1500,  '메로나': 1000,  '팥빙수':2700, '아맛나':1000}  

In [113]:
icecream = {'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000}
new_product = {'팥빙수':2700, '아맛나':1000}
icecream.update(new_product)
print(icecream)

{'탱크보이': 1200, '폴라포': 1200, '빵빠레': 1800, '월드콘': 1500, '메로나': 1000, '팥빙수': 2700, '아맛나': 1000}


### 099. zip과 dict  
아래 두 개의 튜플을 하나의 딕셔너리로 변환하라. keys를 키로, vals를 값으로 result 이름의 딕셔너리로 저장한다.  
```python
keys = ("apple", "pear", "peach")  
vals = (300, 250, 400)  
```
실행 예:  
```python
print(result)  
```
> {'apple': 300, 'pear': 250, 'peach': 400}  

In [114]:
keys = ("apple", "pear", "peach")
vals = (300, 250, 400)
result = dict(zip(keys, vals))
print(result)

{'apple': 300, 'pear': 250, 'peach': 400}


### 100. zip과 dict  
date와 close_price 두 개의 리스트를 close_table 이름의 딕셔너리로 생성하라.  
```python
date = ['09/05', '09/06', '09/07', '09/08', '09/09']  
close_price = [10500, 10300, 10100, 10800, 11000]  
```
실행 예:  
```python
print(close_table)  
```
> {'09/05': 10500, '09/06': 10300, '09/07': 10100, '09/08': 10800, '09/09': 11000}  

In [115]:
date = ['09/05', '09/06', '09/07', '09/08', '09/09']
close_price = [10500, 10300, 10100, 10800, 11000]
close_table = dict(zip(date, close_price))
print(close_table)

{'09/05': 10500, '09/06': 10300, '09/07': 10100, '09/08': 10800, '09/09': 11000}


---
## 101~130. 파이썬 분기문

### 101. bool 타입
파이썬에서 True 혹은 False를 갖는 데이터 타입은 무엇인가?

In [2]:
# bool 타입
print(type(True), type(False))

<class 'bool'> <class 'bool'>


### 102. 비교
아래 코드의 출력 결과를 예상하라  
```python
print(3 == 5)  
```

In [3]:
# False
print(3 == 5)

False


### 103. 비교  
아래 코드의 출력 결과를 예상하라  
```python
print(3 < 5)  
```

In [4]:
# True
print(3 < 5)

True


### 104. 비교
아래 코드의 결과를 예상하라.
```python
x = 4  
print(1 < x < 5)  
```

In [6]:
# True
x = 4
print(1 < x < 5)

True


### 105. 비교  
아래코드의 결과를 예상하라.  
```python
print ((3 == 3) and (4 != 3))  
```

In [7]:
# True
print((3 == 3) and (4 != 3))

True


### 106. 비교  
아래 코드에서 에러가 발생하는 원인에 대해 설명하라.  
```python
print(3 => 4)  
```

In [8]:
# =>연산자는 별도로 존재, 조건식은 3 >= 4 로 표현
print(3 => 4)

SyntaxError: expression cannot contain assignment, perhaps you meant "=="? (2180744063.py, line 2)

### 107. 조건문
아래 코드의 출력 결과를 예상하라  
```python
if 4 < 3:  
    print("Hello World")  
```

In [11]:
# 출력 없음
if 4 < 3: 
    print("Hello World")

### 108. 조건문
아래 코드의 출력 결과를 예상하라  
```python
if 4 < 3:
    print("Hello World.")
else:
    print("Hi, there.")
```

In [12]:
# Hi, there.
if 4 < 3:
    print("Hello World.")
else:
    print("Hi, there.")

Hi, there.


### 109. 조건문
아래 코드의 출력 결과를 예상하라  
```python
if True :
    print ("1")
    print ("2")
else :
    print("3")
print("4")
```

In [13]:
# 1
# 2
# 4
if True:
    print("1")
    print("2")
else:
    print("3")
print("4")

1
2
4


### 110. 조건문
아래 코드의 출력 결과를 예상하라
```python
if True :
    if False:
        print("1")
        print("2")
    else:
        print("3")
else :
    print("4")
print("5")
```

In [15]:
# 3
# 5
if True:
    if False:
        print("1")
        print("2")
    else:
        print("3")
else:
    print("4")
print("5")

3
5


### 111. 사용자 입력
사용자로부터 입력받은 문자열을 두 번 출력하라. 아래는 사용자가 "안녕하세요"를 입력한 경우의 출력 결과이다.
> 안녕하세요  

> 안녕하세요안녕하세요

In [20]:
data = input('입력: ')
print('안녕하세요' * 2)

입력:  안녕하세요


안녕하세요안녕하세요


### 112. 사용자 입력
사용자로부터 하나의 숫자를 입력받고, 입력 받은 숫자에 10을 더해 출력하라.
> 숫자를 입력하세요: 30  

> 40  

In [21]:
data = int(input('숫자 입력: '))
print(data + 10)

숫자 입력:  30


40


### 113. 사용자 입력
사용자로부터 하나의 숫자를 입력 받고 짝수/홀수를 판별하라.
> 30  

> 짝수  

In [23]:
data = int(input('숫자 입력: '))
print(data % 2 and '홀수' or '짝수')

숫자 입력:  12


짝수


### 114. 사용자 입력  
사용자로부터 값을 입력받은 후 해당 값에 20을 더한 값을 출력하라. 단 사용자가 입력한 값과 20을 더한 계산 값이 255를 초과하는 경우 255를 출력해야 한다.
> 입력값: 200  
> 출력값: 220  

> 입력값: 240  
> 출력값: 255  

In [26]:
data = int(input('숫자 입력: ')) + 20
print(255 if data>255 else data)

숫자 입력:  240


255


### 115. 사용자 입력
사용자로부터 하나의 값을 입력받은 후 해당 값에 20을 뺀 값을 출력하라. 단 출력 값의 범위는 0~255이다. 예를 들어 결괏값이 0보다 작은 값이되는 경우 0을 출력하고 255보다 큰 값이 되는 경우 255를 출력해야 한다.
> 입력값: 200  
> 출력값: 180

> 입력값: 15  
> 출력값: 0  

In [35]:
data = int(input('숫자 입력: ')) - 20
print(0 if data<0 else 255 if data>255 else data)

숫자 입력:  15


0


### 116. 사용자 입력
사용자로부터 입력 받은 시간이 정각인지 판별하라.
> 현재시간:02:00  
> 정각 입니다.  

> 현재시간:03:10  
> 정각이 아닙니다  

In [40]:
data = input('시간 입력 (입력예 10:25) : ')
print(int(data.split(':')[1]) and '정각이 아닙니다.' or '정각 입니다.')

시간 입력 (입력예 10:25) :  03:10


정각이 아닙니다.


### 117. 사용자 입력  
사용자로 입력받은 단어가 아래 fruit 리스트에 포함되어 있는지를 확인하라. 포함되었다면 "정답입니다"를 아닐 경우 "오답입니다" 출력하라.
```python
fruit = ["사과", "포도", "홍시"]
```
> 좋아하는 과일은? 사과  
> 정답입니다.

In [45]:
fruit = ["사과", "포도", "홍시"]
data = input('좋아하는 과일은? ')
print('정답입니다.' if data in fruit else '오답입니다.')

좋아하는 과일은?  사과


정답입니다.


### 118. 사용자 입력
투자 경고 종목 리스트가 있을 때 사용자로부터 종목명을 입력 받은 후 해당 종목이 투자 경고 종목이라면 '투자 경고 종목입니다'를 아니면 "투자 경고 종목이 아닙니다."를 출력하는 프로그램을 작성하라.
```python
warn_investment_list = ["Microsoft", "Google", "Naver", "Kakao", "SAMSUNG", "LG"]
```

In [47]:
warn_investment_list = ["Microsoft", "Google", "Naver", "Kakao", "SAMSUNG", "LG"]
data = input('종목명: ')
print(data in warn_investment_list and '투자 경고 종목입니다.' or '투자 경고 종목이 아닙니다.')

종목명:  Kakao


투자 경고 종목입니다.


### 119. 사용자 입력
아래와 같이 fruit 딕셔너리가 정의되어 있다. 사용자가 입력한 값이 딕셔너리 키 (key) 값에 포함되었다면 "정답입니다"를 아닐 경우 "오답입니다" 출력하라.
```python
fruit = {"봄" : "딸기", "여름" : "토마토", "가을" : "사과"}
```
> 제가좋아하는계절은: 봄  
> 정답입니다.  

In [51]:
fruit = {"봄" : "딸기", "여름" : "토마토", "가을" : "사과"}
data = input('제가 좋아하는 계절은? ')
print(data in fruit.keys() and '정답입니다.' or '오답입니다.')

제가 좋아하는 계절은?  여름


정답입니다.


### 120. 사용자 입력
아래와 같이 fruit 딕셔너리가 정의되어 있다. 사용자가 입력한 값이 딕셔너리 값 (value)에 포함되었다면 "정답입니다"를 아닐 경우 "오답입니다" 출력하라.
```python
fruit = {"봄" : "딸기", "여름" : "토마토", "가을" : "사과"}
```
> 좋아하는과일은? 한라봉  
> 오답입니다.  

In [53]:
fruit = {"봄" : "딸기", "여름" : "토마토", "가을" : "사과"}
data = input('좋아하는 과일은? ')
print(data in fruit.values() and '정답입니다.' or '오답입니다.')

좋아하는 과일은?  딸기


정답입니다.


### 121. 사용자 입력
사용자로부터 문자 한 개를 입력 받고, 소문자일 경우 대문자로, 대문자 일 경우, 소문자로 변경해서 출력하라.
> a  
> A

힌트-1 : islower() 함수는 문자의 소문자 여부를 판별합니다. 만약 소문자일 경우 True, 대문자일 경우 False를 반환합니다.  
힌트-2 : upper() 함수는 대문자로, lower() 함수는 소문자로 변경합니다.

In [57]:
data = input()
print(data.islower() and data.upper() or data.lower())

 a


A


### 122. 사용자 입력
점수 구간에 해당하는 학점이 아래와 같이 정의되어 있다. 사용자로부터 score를 입력받아 학점을 출력하라.

|점수|학점|
|--|--|
|81~100|A|
|61~80|B|
|41~60|C|
|21~40|D|
|0~20|E|
> score: 83  
> grade is A  

In [67]:
data = int(input('score: '))
print(f"grade is {'EDCBA'[(data-1)//20]}")

score:  81


grade is A


### 123. 사용자 입력
사용자로부터 달러, 엔, 유로, 또는 위안 금액을 입력받은 후 이를 원으로 변환하는 프로그램을 작성하라. 각 통화별 환율은 다음과 같다. 사용자는 100 달러, 1000 엔, 13 유로, 100 위안과 같이 금액과 통화명 사이에 공백을 넣어 입력한다고 가정한다.

|통화명|환율|
|--|--|
|달러|1167|
|엔|1.096|
|유로|1268|
|위안|171|
> 입력: 100 달러  
> 116700.00 원  

In [15]:
data = input('입력 (100 달러): ').split()
ex_rate = { '달러': 1167, '엔': 1.096, '유로': 1268, '위안': 171 }
print(f"{float(data[0]) * ex_rate[data[1]]:.2f} 원")

입력 (100 달러):  100 달러


116700.00 원


### 124. 사용자 입력
사용자로부터 세 개의 숫자를 입력 받은 후 가장 큰 숫자를 출력하라.

> input number1: 10   
> input number2: 9  
> input number3: 20  
> 20

In [18]:
data = [int(input(f"input number{i+1}: ")) for i in range(3)]
print(max(data))

input number1:  10
input number2:  9
input number3:  20


20


### 125. 사용자 입력
휴대폰 번호 앞자리에 따라 통신사는 아래와 같이 구분된다. 사용자로부터 휴대전화 번호를 입력 받고, 통신사를 출력하는 프로그램을 작성하라.

|번호|통신사|
|--|--|
|011|SKT|
|016|KT|
|019|LGU|
|010|알수없음|
> 휴대전화 번호 입력: 011-345-1922  
> 당신은 SKT 사용자입니다.  

In [21]:
data = input('휴대전화 번호 입력: ')
phoneInfo = { '011': 'SKT', '016': 'KT', '019': 'LGU', '010': '알수없음' }
print(f"당신은 {phoneInfo[data.split('-')[0]]} 사용자입니다.")

휴대전화 번호 입력:  011-345-1922


당신은 SKT 사용자입니다.


### 126. 사용자 입력
우편번호는 5자리로 구성되는데, 앞의 세자리는 구를 나타낸다. 예를들어, 강북구의 경우 010, 011, 012 세 자리로 시작한다.

|-|0|1|2|3|4|5|6|7|8|9|
|--|--|--|--|--|--|--|--|--|--|--|
|01|강북구|강북구|강북구|도봉구|도봉구|도봉구|노원구|노원구|노원구|노원구|

사용자로 부터 5자리 우편번호를 입력받고 구를 판별하라

> 우편번호: 01400  
> 도봉구  

In [30]:
data = input('우편번호: ')
if data[:2] == '01':
    num = int(data[2])
    print(num<3 and '강북구' or num<6 and '도봉구' or '노원구')
else:
    print('알수없음')

우편번호:  01400


도봉구


### 127.  사용자 입력
주민등록번호 뒷 자리 7자리 중 첫째 자리는 성별을 나타내는데, 1, 3은 남자 2, 4는 여자를 의미한다. 사용자로부터 13자리의 주민등록번호를 입력 받은 후 성별 (남자, 여자)를 출력하는 프로그램을 작성하라.

> 주민등록번호: 821010-1635210  
> 남자  

In [53]:
data = input('주민등록번호: ').split('-')[1][0]
print(data in '13' and '남자' or data in '24' and '여자' or '모름')

주민등록번호:  821010-1635210


남자


### 128. 사용자 입력
주민등록번호의 뒷 자리 7자리 중 두번째와 세번째는 지역코드를 의미한다. 주민 등록 번호를 입력 받은 후 출생지가 서울인지 아닌지 판단하는 코드를 작성하라

|지역코드|출생지|
|--|--|
|00 ~ 08|서울|
|09 ~ 12|부산|

> 주민등록번호: 821010-1635210  
>서울이 아닙니다.  

> 주민등록번호: 861010-1015210  
> 서울 입니다.  

In [60]:
data = input('주민등록번호: ').split('-')[1][1:3]
print(int(data) <= 8 and '서울 입니다.' or '서울이 아닙니다.')

주민등록번호:  861010-1015210


서울 입니다.


### 129. 사용자 입력  
주민등록번호는 13자리로 구성되는데 마지막 자리수는 주민등록번호의 유효성을 체크하는데 사용된다.  
먼저 앞에서부터 12자리의 숫자에 2, 3, 4, 5, 6, 7, 8, 9, 2, 3, 4, 5를 차례로 곱한 뒤 그 값을 전부 더한다.   
연산 결과 값을 11로 나누면 나머지가 나오는데 11에서 나머지를 뺀 값이 주민등록번호의 마지막 번호가 된다.  

```text
  8 2 1 0 1 0 - 1 6 3 5 2 1 0
x 2 3 4 5 6 7   8 9 2 3 4 5 
-----------------------------
1차 계산: (8*2 + 2*3 + 1*4 + 0*5 + 1*6 + 0*7 + 1*8 + 6*9 + 3*2 + 5*3 + 2*4 + 1*5) = (128 % 11) = 7
2차 계산: 11 -7 = 4
```

위와 같이 821010-1635210에 대해서 계산을 해보면 마지막 자리는 4가 되어야 함을 알 수 있다. 즉, 821010-1635210은 유효하지 않은 주민등록번호임을 알 수 있다.

다음과 같이 사용자로부터 주민등록번호를 입력받은 후 주민등록번호가 유효한지를 출력하는 프로그램을 작성하라.

> 주민등록번호: 821010-1635210  
> 유효하지 않은 주민등록번호입니다.  

In [80]:
data = input('주민등록번호: ').replace('-', '')
s = sum([int(x)*int(y) for x, y in zip(data, '234567892345')]) 
print(int(data[-1]) == 11-(s%11) and '유효한 주민등록번호 입니다.' or '유효하지 않은 주민등록번호 입니다.')

주민등록번호:  821010-1635210


유효하지 않은 주민등록번호 입니다.


### 130. json 데이터
아래 코드는 비트코인의 가격 정보를 딕셔너리로 가져오는 코드이다.
```python
import requests
btc = requests.get("https://api.bithumb.com/public/ticker/").json()['data']
```
btc 딕셔너리 안에는 시가, 종가, 최고가, 최저가 등이 저장되어 있다. 최고가와 최저가의 차이를 변동폭으로 정의할 때 (시가 + 변동폭)이 종가 보다 높을 경우 "상승장", 그렇지 않은 경우 "하락장" 문자열을 출력하라.

|Key Name|Description|
|--|--|
|opening_price|최근 24시간 내 시작 거래금액|
|closing_price|최근 24시간 내 마지막 거래금액|
|min_price|최근 24시간 내 최저 거래금액|
|max_price|최근 24시간 내 최고 거래금액|

In [84]:
import requests
btc = requests.get("https://api.bithumb.com/public/ticker/").json()['data']
flag = float(btc['opening_price']) + float(btc['max_price']) - float(btc['min_price']) - float(btc['closing_price'])
print(flag > 0 and '상승장' or flag < 0 and '하락장' or '보합')

상승장


---
## 131~200. 파이썬 반복문

### 131. for문
for문의 실행결과를 예측하라.
```python
과일 = ["사과", "귤", "수박"]
for 변수 in 과일:
    print(변수)
```

In [1]:
# 사과
# 귤
# 수박
과일 = ["사과", "귤", "수박"]
for 변수 in 과일:
    print(변수)

사과
귤
수박


### 132. for문
for문의 실행결과를 예측하라.
```python
과일 = ["사과", "귤", "수박"]
for 변수 in 과일:  print("#####")
```

In [3]:
# #####
# #####
# #####
과일 = ["사과", "귤", "수박"]
for 변수 in 과일:
  print("#####")

#####
#####
#####


### 133. for문
다음 for 문과 동일한 기능을 수행하는 코드를 작성하세요.
```python
for 변수 in ["A", "B", "C"]:
    print(변수)
```

In [7]:
_ = [print(x) for x in 'ABC']

A
B
C


### 134. for문
for문을 풀어서 동일한 동작을하는 코드를 작성하라.
```python
for 변수 in ["A", "B", "C"]:
  print("출력:", 변수)
```

In [10]:
_ = [print('출력:', x) for x in 'ABC']

출력: A
출력: B
출력: C


### 135. for문
for문을 풀어서 동일한 동작을 하는 코드를 작성하라.
```python
for 변수 in ["A", "B", "C"]:
  b = 변수.lower()
  print("변환:", b)
```

In [11]:
_ = [print('변환:', x.lower()) for x in 'ABC']

변환: a
변환: b
변환: c


### 136. for문
다음 코드를 for문으로 작성하라.
```python
변수 = 10
print(변수)
변수 = 20
print(변수)
변수 = 30
print(변수)
```

In [12]:
for 변수 in [10,20,30]:
    print(변수)

10
20
30


### 137. for문
다음 코드를 for문으로 작성하라.
```python
print(10)
print(20)
print(30)
```

In [13]:
_ = [print(x) for x in range(10,31,10)]

10
20
30


### 138. for문
다음 코드를 for문으로 작성하라.
```python
print(10)
print("-------")
print(20)
print("-------")
print(30)
print("-------")
```

In [14]:
for x in range(10,31,10):
    print(x)
    print('-'*7)

10
-------
20
-------
30
-------


### 139. for문
다음 코드를 for문으로 작성하라.
```python
print("++++")
print(10)
print(20)
print(30)
```

In [15]:
print('+' * 4)
_ = [print(x) for x in range(10,31,10)]

++++
10
20
30


### 140. for문
다음 코드를 for문으로 작성하라.
```python
print("-------")
print("-------")
print("-------")
print("-------")
```

In [18]:
_ = [print('-' * 7) for _ in range(4)]

-------
-------
-------
-------


### 141. for문
다음과 같이 판매가가 저장된 리스트가 있을 때 부가세가 포함된 가격을 for 문을 사용해서 화면에 출력하라. 단 부가세는 10원으로 가정한다.
```python
리스트 = [100, 200, 300]
```
> 110  
> 210  
> 310  

In [19]:
리스트 = [100, 200, 300]
_ = [print(x + 10) for x in 리스트]

110
210
310


### 142. for문
for 문을 사용해서 리스트에 저장된 값을 다음과 같이 출력하라.
```python
리스트 = ["김밥", "라면", "튀김"]
```
> 오늘의 메뉴: 김밥  
> 오늘의 메뉴: 라면  
> 오늘의 메뉴: 튀김  

In [21]:
리스트 = ["김밥", "라면", "튀김"]
_ = [print('오늘의 메뉴:', x) for x in 리스트]

오늘의 메뉴: 김밥
오늘의 메뉴: 라면
오늘의 메뉴: 튀김


### 143. for문
리스트에 주식 종목이름이 저장돼 있다.
```python
리스트 = ["SK하이닉스", "삼성전자", "LG전자"]
```
저장된 문자열의 길이를 다음과 같이 출력하라.  
> 6  
> 4  
> 4  

In [22]:
리스트 = ["SK하이닉스", "삼성전자", "LG전자"]
_ = [print(len(x)) for x in 리스트]

6
4
4


### 144. for문
리스트에는 동물이름이 문자열로 저장돼 있다.
```python
리스트 = ['dog', 'cat', 'parrot']
```
동물 이름과 글자수를 다음과 같이 출력하라.  
> dog 3  
> cat 3  
> parrot 6  

In [23]:
리스트 = ['dog', 'cat', 'parrot']
_ = [print(x, len(x)) for x in 리스트]

dog 3
cat 3
parrot 6


### 145. for문
리스트에 동물 이름 저장돼 있다.
```python
리스트 = ['dog', 'cat', 'parrot']
```
for문을 사용해서 동물 이름의 첫 글자만 출력하라.  
> d  
> c  
> p  

In [24]:
리스트 = ['dog', 'cat', 'parrot']
_ = [print(x[0]) for x in 리스트]

d
c
p


### 146. for문
리스트에는 세 개의 숫자가 바인딩돼 있다.
```python
리스트 = [1, 2, 3]
```
for문을 사용해서 다음과 같이 출력하라.  
> 3 x 1  
> 3 x 2  
> 3 x 3  

In [26]:
리스트 = [1, 2, 3]
_ = [print(f"3 x {x}") for x in 리스트]

3 x 1
3 x 2
3 x 3


### 147. for문
리스트에는 세 개의 숫자가 바인딩돼 있다.
```python
리스트 = [1, 2, 3]
```
for문을 사용해서 다음과 같이 출력하라.  
> 3 x 1 = 3  
> 3 x 2 = 6  
> 3 x 3 = 9  

In [27]:
리스트 = [1, 2, 3]
_ = [print(f"3 x {x} = {3*x}") for x in 리스트]

3 x 1 = 3
3 x 2 = 6
3 x 3 = 9


### 148. for문
리스트에는 네 개의 문자열이 바인딩돼 있다.
```python
리스트 = ["가", "나", "다", "라"]
```
for문을 사용해서 다음과 같이 출력하라.  
> 나  
> 다  
> 라  

In [28]:
리스트 = ["가", "나", "다", "라"]
_ = [print(x) for x in 리스트[1:]]

나
다
라


### 149. for문
리스트에는 네 개의 문자열이 바인딩돼 있다.
```python
리스트 = ["가", "나", "다", "라"]
```
for문을 사용해서 다음과 같이 출력하라.  
> 가  
> 다  

In [29]:
리스트 = ["가", "나", "다", "라"]
_ = [print(x) for x in 리스트[::2]]

가
다


### 150. for문
리스트에는 네 개의 문자열이 바인딩돼 있다.
```python
리스트 = ["가", "나", "다", "라"]
```
for문을 사용해서 다음과 같이 출력하라.  
> 라  
> 다  
> 나  
> 가  

In [30]:
리스트 = ["가", "나", "다", "라"]
_ = [print(x) for x in 리스트[::-1]]

라
다
나
가


### 151. for문
리스트에는 네 개의 정수가 저장돼 있다.
```python
리스트 = [3, -20, -3, 44]
```
for문을 사용해서 리스트의 음수를 출력하라.
> -20  
> -3  

In [31]:
리스트 = [3, -20, -3, 44]
_ = [print(x) for x in 리스트 if x<0]

-20
-3


### 152. for문
for문을 사용해서 3의 배수만을 출력하라.
```python
리스트 = [3, 100, 23, 44]
```
> 3  

In [33]:
리스트 = [3, 100, 23, 44]
_ = [print(x) for x in 리스트 if not(x % 3)]

3


### 153. for문
리스트에서 20 보다 작은 3의 배수를 출력하라
```python
리스트 = [13, 21, 12, 14, 30, 18]
```
> 12  
> 18  

In [34]:
리스트 = [13, 21, 12, 14, 30, 18]
_ = [print(x) for x in 리스트 if x < 20 and not(x % 3)]

12
18


### 154. for문
리스트에서 세 글자 이상의 문자를 화면에 출력하라
```python
리스트 = ["I", "study", "python", "language", "!"]
```
> study  
> python  
> language  

In [35]:
리스트 = ["I", "study", "python", "language", "!"]
_ = [print(x) for x in 리스트 if len(x) >= 3]

study
python
language


### 155. for문
리스트에서 대문자만 화면에 출력하라.
```python
리스트 = ["A", "b", "c", "D"]
```
> A  
> D  

In [36]:
리스트 = ["A", "b", "c", "D"]
_ = [print(x) for x in 리스트 if x.isupper()]

A
D


### 156. for문
리스트에서 소문자만 화면에 출력하라.
```python
리스트 = ["A", "b", "c", "D"]
```
> b  
> c  

In [37]:
리스트 = ["A", "b", "c", "D"]
_ = [print(x) for x in 리스트 if x.islower()]

b
c


### 157. for문
이름의 첫 글자를 대문자로 변경해서 출력하라.
```python
리스트 = ['dog', 'cat', 'parrot']
```
> Dog  
> Cat  
> Parrot

(참고) upper() 메서드는 문자열을 대문자로 변경합니다.
```python
변수 = "a"
a.upper()
```
> A  
```python
변수 = "abc"
변수.upper()
```
> ABC  

In [38]:
리스트 = ['dog', 'cat', 'parrot']
_ = [print(x.capitalize()) for x in 리스트]

Dog
Cat
Parrot


### 158. for문
파일 이름이 저장된 리스트에서 확장자를 제거하고 파일 이름만 화면에 출력하라. (힌트: split() 메서드)
```python
리스트 = ['hello.py', 'ex01.py', 'intro.hwp']
```
> hello  
> ex01  
> intro  

In [39]:
리스트 = ['hello.py', 'ex01.py', 'intro.hwp']
_ = [print(x.split('.')[0]) for x in 리스트]

hello
ex01
intro


### 159. for문
파일 이름이 저장된 리스트에서 확장자가 .h인 파일 이름을 출력하라.
```python
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
```
> intra.h  
> define.h  

In [40]:
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
_ = [print(x) for x in 리스트 if x.split('.')[-1] == 'h']

intra.h
define.h


### 160. for문
파일 이름이 저장된 리스트에서 확장자가 .h나 .c인 파일을 화면에 출력하라.
```python
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
```
> intra.h  
> intra.c  
> define.h  

In [41]:
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
_ = [print(x) for x in 리스트 if x.split('.')[-1] in 'hc']

intra.h
intra.c
define.h


### 161. for문
for문과 range 구문을 사용해서 0~99까지 한 라인에 하나씩 순차적으로 출력하는 프로그램을 작성하라.

In [42]:
_ = [print(x) for x in range(100)]

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


### 162. for문
월드컵은 4년에 한 번 개최된다. range()를 사용하여 2002~2050년까지 중 월드컵이 개최되는 연도를 출력하라.

> 2002  
> 2006  
> 2010  
> ...  
> 2042  
> 2046  
> 2050

참고) range의 세번 째 파라미터는 증감폭을 결정합니다.
```python
print(list(range(0, 10, 2)))
```
> [0, 2, 4, 6, 8]  

In [43]:
_ = [print(x) for x in range(2002,2051,4)]

2002
2006
2010
2014
2018
2022
2026
2030
2034
2038
2042
2046
2050


### 163. for문
1부터 30까지의 숫자 중 3의 배수를 출력하라.
> 3  
> 6  
> 9  
> 12  
> 15  
> 18  
> 21  
> 24  
> 27  
> 30 

In [46]:
_ = [print(x) for x in range(3,31,3)]

3
6
9
12
15
18
21
24
27
30


### 164. for문
99부터 0까지 1씩 감소하는 숫자들을, 한 라인에 하나씩 출력하라.

In [47]:
_ = [print(x) for x in range(99,-1,-1)]

99
98
97
96
95
94
93
92
91
90
89
88
87
86
85
84
83
82
81
80
79
78
77
76
75
74
73
72
71
70
69
68
67
66
65
64
63
62
61
60
59
58
57
56
55
54
53
52
51
50
49
48
47
46
45
44
43
42
41
40
39
38
37
36
35
34
33
32
31
30
29
28
27
26
25
24
23
22
21
20
19
18
17
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1
0


### 165. for문
for문을 사용해서 아래와 같이 출력하라.
> 0.0  
> 0.1  
> 0.2  
> 0.3  
> 0.4  
> 0.5  
> ...  
> 0.9  

In [50]:
_ = [print(x/10) for x in range(10)]

0.0
0.1
0.2
0.3
0.4
0.5
0.6
0.7
0.8
0.9


### 166. for문
구구단 3단을 출력하라.
> 3x1 = 3  
> 3x2 = 6  
> 3x3 = 9  
> 3x4 = 12  
> 3x5 = 15  
> 3x6 = 18  
> 3x7 = 21  
> 3x8 = 24  
> 3x9 = 27  

In [52]:
_ = [print(f"3x{x} = {3*x}") for x in range(1,10)]

3x1 = 3
3x2 = 6
3x3 = 9
3x4 = 12
3x5 = 15
3x6 = 18
3x7 = 21
3x8 = 24
3x9 = 27


### 167. for문
구구단 3단을 출력하라. 단 홀수 번째만 출력한다.
> 3x1 = 3  
> 3x3 = 9  
> 3x5 = 15  
> 3x7 = 21  
> 3x9 = 27  

In [54]:
_ = [print(f"3x{x} = {3*x}") for x in range(1,10,2)]

3x1 = 3
3x3 = 9
3x5 = 15
3x7 = 21
3x9 = 27


### 168. for문
1~10까지의 숫자에 대해 모두 더한 값을 출력하는 프로그램을 for 문을 사용하여 작성하라.
> 합 : 55  

In [61]:
print('합:', sum(range(11)))

합: 55


### 169. for문
1~10까지의 숫자 중 모든 홀수의 합을 출력하는 프로그램을 for 문을 사용하여 작성하라.
> 합: 25  

In [62]:
print('합:',sum(range(1,11,2)))

합: 25


### 170. for문
1~10까지의 숫자를 모두 곱한 값을 출력하는 프로그램을 for 문을 사용하여 작성하라.

In [75]:
r = 1; print([r:=r*x for x in range(1,11)][-1])

3628800


### 171. for문
아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.
```python
price_list = [32100, 32150, 32000, 32500]
```
> 32100  
> 32150  
> 32000  
> 32500  

In [76]:
price_list = [32100, 32150, 32000, 32500]
_ = [print(price_list[i]) for i in range(len(price_list))]

32100
32150
32000
32500


### 172. for문
아래와 같이 리스트의 데이터를 출력하라. 단, for문과 enumerate문을 사용하라.
```python
price_list = [32100, 32150, 32000, 32500]
```
> 0 32100  
> 1 32150  
> 2 32000  
> 3 32500  

In [79]:
price_list = [32100, 32150, 32000, 32500]
_ = [print(i, data) for i, data in enumerate(price_list)]

0 32100
1 32150
2 32000
3 32500


### 173. for문
아래와 같이 리스트의 데이터를 출력하라. 단, for문과 enumerate문을 사용하라.
```python
price_list = [32100, 32150, 32000, 32500]
```
> 3 32100  
> 2 32150  
> 1 32000  
> 0 32500  

In [82]:
price_list = [32100, 32150, 32000, 32500]
_ = [print(len(price_list) - 1 - i, data) for i, data in enumerate(price_list)]

3 32100
2 32150
1 32000
0 32500


### 174. for문
아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.
```python
price_list = [32100, 32150, 32000, 32500]
```
> 100 32150  
> 110 32000  
> 120 32500  

In [85]:
price_list = [32100, 32150, 32000, 32500]
_ = [print(100+10*i, price_list[i+1]) for i in range(3)]

100 32150
110 32000
120 32500


### 175. for문
my_list를 아래와 같이 출력하라.
```python
my_list = ["가", "나", "다", "라"]
```
> 가 나  
> 나 다  
> 다 라  

In [87]:
my_list = ["가", "나", "다", "라"]

_ = [print(my_list[i-1], my_list[i]) for i in range(1,len(my_list))]

가 나
나 다
다 라


### 176. for문
리스트를 아래와 같이 출력하라.
```python
my_list = ["가", "나", "다", "라", "마"]
```
> 가 나 다  
> 나 다 라  
> 다 라 마  

In [89]:
my_list = ["가", "나", "다", "라", "마"]
_ = [print(*my_list[i:i+3]) for i in range(3)]

가 나 다
나 다 라
다 라 마


### 177. for문
반복문과 range 함수를 사용해서 my_list를 아래와 같이 출력하라.
```python
my_list = ["가", "나", "다", "라"]
```
> 라 다  
> 다 나  
> 나 가  

In [93]:
my_list = ["가", "나", "다", "라"]
_ = [print(*my_list[2-i:4-i][::-1]) for i in range(3)]

라 다
다 나
나 가


### 178. for문
리스트에는 네 개의 정수가 저장되어 있다. 각각의 데이터에 대해서 자신과 우측값과의 차분값을 화면에 출력하라.
```python
my_list = [100, 200, 400, 800]
```
예를들어 100을 기준으로 우측에 위치한 200과의 차분 값를 화면에 출력하고, 200을 기준으로 우측에 위치한 400과의 차분값을 화면에 출력한다. 이어서 400을 기준으로 우측에 위치한 800과의 차분값을 화면에 출력한다.
> 100  
> 200  
> 400  

In [95]:
my_list = [100, 200, 400, 800]
_ = [print(my_list[i] - my_list[i-1]) for i in range(1,len(my_list))]

100
200
400


### 179. for문
리스트에는 6일 간의 종가 데이터가 저장되어 있다. 종가 데이터의 3일 이동 평균을 계산하고 이를 화면에 출력하라.
```python
my_list = [100, 200, 400, 800, 1000, 1300]
```
첫 번째 줄에는 100, 200, 400의 평균값이 출력된다. 두 번째 줄에는 200, 400, 800의 평균값이 출력된다. 같은 방식으로 나머지 데이터의 평균을 출력한다.
> 233.33333333333334  
> 466.6666666666667  
> 733.3333333333334  
> 1033.3333333333333  

In [96]:
my_list = [100, 200, 400, 800, 1000, 1300]
_ = [print(sum(my_list[i-2:i+1])/3) for i in range(2,len(my_list))]

233.33333333333334
466.6666666666667
733.3333333333334
1033.3333333333333


### 180. for문
리스트에 5일간의 저가, 고가 정보가 저장돼 있다. 고가와 저가의 차를 변동폭이라고 정의할 때, low, high 두 개의 리스트를 사용해서 5일간의 변동폭을 volatility 리스트에 저장하라.
```python
low_prices  = [100, 200, 400, 800, 1000]
high_prices = [150, 300, 430, 880, 1000]
```

In [100]:
low_prices  = [100, 200, 400, 800, 1000]
high_prices = [150, 300, 430, 880, 1000]
volatility = [(h - l) for l, h in zip(low_prices, high_prices)]
print(volatility)

[50, 100, 30, 80, 0]


### 181. 다중배열
아래 표에서 하나의 행을 하나의 리스트로, 총 3개의 리스트를 갖는 이차원 리스트 apart를 정의하라.
||열1|열2|
|--|--|--|
|행1|101호|102호|
|행2|201호|202호|
|행3|301호|302호|

In [1]:
apart = [
    ['101호', '102호'],
    ['201호', '202호'],
    ['301호', '302호'],
]
print(apart)

[['101호', '102호'], ['201호', '202호'], ['301호', '302호']]


### 182. 다중배열
아래 표에서 하나의 열을 하나의 리스트로, 총 2개의 리스트를 갖는 이차원 리스트 stock을 정의하라.

|시가|종가|
|--|--|
|100|80|
|200|210|
|300|330|

In [3]:
stock = [
    ['시가', 100, 200, 300],
    ['종가', 80, 210, 330],
]
print(stock)

[['시가', 100, 200, 300], ['종가', 80, 210, 330]]


### 183. 다중배열
아래 표를 stock 이름의 딕셔너리로 표현하라.시가를 key로 저장하고, 나머지 같은 열의 데이터를 리스트로 저장해서 value로 저장한다. 종가 역시 key로 저장하고 나머지 같은 열의 데이터를 리스트로 저장해서 value로 저장한다.

|시가|종가|
|--|--|
|100|80|
|200|210|
|300|330|

In [4]:
stock = {
    '시가': [100,200,300],
    '종가': [80,210,330],
}
print(stock)

{'시가': [100, 200, 300], '종가': [80, 210, 330]}


### 184. 다중배열
아래 표를 stock 이라는 이름의 딕셔너리로 표현하라. 날짜를 key로 저장하고, 나머지 같은 행의 데이터를 리스트로 저장해서 value로 저장한다. 첫 열이 날짜이다.
|날짜|열1|열2|열3|열4|
|--|--|--|--|--|
|10/10|80|110|70|90|
|10/11|210|230|190|200|

In [5]:
stock = {
    '10/10': [80,110,70,90],
    '10/11': [210,230,190,200],
}
print(stock)

{'10/10': [80, 110, 70, 90], '10/11': [210, 230, 190, 200]}


### 185. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 101 호  
> 102 호  
> 201 호  
> 202 호  
> 301 호  
> 302 호  

In [8]:
apart = [ [101, 102], [201, 202], [301, 302] ]
_ = [print(y, '호') for x in apart for y in x]

101 호
102 호
201 호
202 호
301 호
302 호


### 186. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 301 호  
> 302 호  
> 201 호  
> 202 호  
> 101 호  
> 102 호  

In [10]:
apart = [ [101, 102], [201, 202], [301, 302] ]
_ = [print(y, '호') for x in apart[::-1] for y in x]

301 호
302 호
201 호
202 호
101 호
102 호


### 187. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 302 호  
> 301 호  
> 202 호  
> 201 호  
> 102 호  
> 101 호  

In [11]:
apart = [ [101, 102], [201, 202], [301, 302] ]
_ = [print(y, '호') for x in apart[::-1] for y in x[::-1]]

302 호
301 호
202 호
201 호
102 호
101 호


### 188. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 101 호  
> <pre>-----</pre>
> 102 호  
> <pre>-----</pre>
> 201 호  
> <pre>-----</pre>
> 202 호  
> <pre>-----</pre>
> 301 호  
> <pre>-----</pre>
> 302 호  
> <pre>-----</pre>

In [12]:
apart = [ [101, 102], [201, 202], [301, 302] ]
_ = [print(y, '호', '\n-----') for x in apart for y in x]

101 호 
-----
102 호 
-----
201 호 
-----
202 호 
-----
301 호 
-----
302 호 
-----


### 189. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 101 호  
> 102 호  
> <pre>-----</pre>
> 201 호  
> 202 호  
> <pre>-----</pre>
> 301 호  
> 302 호  
> <pre>-----</pre>

In [51]:
apart = [ [101, 102], [201, 202], [301, 302] ]
# _ = [print('\n'.join(map(lambda i:f"{i} 호", x)), '\n-----') for x in apart]
_ = [print(y,'호', '\n-----' if str(y).endswith('2') else '') for x in apart for y in x]

101 호 
102 호 
-----
201 호 
202 호 
-----
301 호 
302 호 
-----


### 190. 다중배열
리스트에 저장된 데이터를 아래와 같이 출력하라.
```python
apart = [ [101, 102], [201, 202], [301, 302] ]
```
> 101 호  
> 102 호  
> 201 호  
> 202 호  
> 301 호  
> 302 호  
> <pre>-----</pre>

In [52]:
apart = [ [101, 102], [201, 202], [301, 302] ]
_ = [print(y, '호') for x in apart for y in x]
print('-'*5)

101 호
102 호
201 호
202 호
301 호
302 호
-----


### 191. 다중배열
data에는 매수한 종목들의 OHLC (open/high/low/close) 가격 정보가 바인딩 되어있다.
```python
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]
```
수수료를 0.014 %로 가정할 때, 각 가격에 수수료를 포함한 가격을 한라인에 하나씩 출력하라.
> 2000.28  
> 3050.427  
> 2050.2870000000003  
> ...

In [53]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]
_ = [print(y*1.00014) for x in data for y in x] 

2000.28
3050.427
2050.2870000000003
1980.2772
7501.05
2050.2870000000003
2050.2870000000003
1980.2772
15452.163
15052.107
15552.177
14902.086000000001


### 192. 다중배열
191번의 출력 결과에 행단위로 "----" 구분자를 추가하라.
> 2000.28  
> 3050.427  
> 2050.2870000000003  
> 1980.2772  
> <pre>----</pre>
> 7501.05  
> 2050.2870000000003  
> 2050.2870000000003  
> 1980.2772  
> <pre>----</pre>
> 15452.163  
> 15052.107  
> 15552.177  
> 14902.086000000001  
> <pre>----</pre>

In [54]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]
_ = [print(y*1.00014, '\n----' if i==len(x)-1 else '') for x in data for i,y in enumerate(x)] 

2000.28 
3050.427 
2050.2870000000003 
1980.2772 
----
7501.05 
2050.2870000000003 
2050.2870000000003 
1980.2772 
----
15452.163 
15052.107 
15552.177 
14902.086000000001 
----


### 193. 다중배열
192 번 문제의 결괏값을 result 이름의 리스트에 1차원 배열로 저장하라.
```python
print(result)
```
> [2000.28, 3050.427, 2050.2870000000003, 1980.2772, 7501.05, 2050.2870000000003, 2050.2870000000003, ...]  

In [55]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]
result = [y*1.00014 for x in data for i,y in enumerate(x)] 
print(result)

[2000.28, 3050.427, 2050.2870000000003, 1980.2772, 7501.05, 2050.2870000000003, 2050.2870000000003, 1980.2772, 15452.163, 15052.107, 15552.177, 14902.086000000001]


### 194. 다중배열
191번 문제의 결괏값을 result 이름의 리스트에 2차원 배열로 저장하라. 저장 포맷은 아래와 같다. 각 행에 대한 데이터끼리 리스트에 저장되어야 한다.
```python
print(result)
```
> [  
>   [2000.28, 3050.427, 2050.2870000000003, 1980.2772],  
>   [7501.05, 2050.2870000000003, 2050.2870000000003, 1980.2772],  
>   [15452.163, 15052.107, 15552.177, 14902.086000000001]  
> ]  

In [56]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]
result = [[y*1.00014 for y in x] for x in data]
print(result)

[[2000.28, 3050.427, 2050.2870000000003, 1980.2772], [7501.05, 2050.2870000000003, 2050.2870000000003, 1980.2772], [15452.163, 15052.107, 15552.177, 14902.086000000001]]


### 195. 다중배열
ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 화면에 종가데이터를 출력하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
> 100  
> 190  
> 310  

In [58]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
_ = [print(x[-1]) for x in ohlc[1:]]

100
190
310


### 196. 다중배열
ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 종가가 150원보다 큰경우에만 종가를 출력하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
> 190  
> 310  

In [59]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
_ = [print(x[-1]) for x in ohlc[1:] if x[-1]>150]

190
310


### 197. 다중배열
ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 종가가 시가 보다 크거나 같은 경우에만 종가를 출력하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
> 100  
> 310  

In [60]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
_ = [print(x[-1]) for x in ohlc[1:] if x[-1]>=x[0]]

100
310


### 198. 다중배열
ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 고가와 저가의 차이를 변동폭으로 정의할 때 변동폭을 volatility 이름의 리스트에 저장하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
```python
print(volatility)
```
> [40, 30, 10]  

In [63]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
volatility = [x[1]-x[2] for x in ohlc[1:]]
print(volatility)

[40, 30, 10]


### 199. 다중배열
리스트에는 3일 간의 ohlc 데이터가 저장돼 있다. 종가가 시가보다 높은 날의 변동성 (고가 - 저가)을 화면에 출력하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
종가가 시가보다 높은 거래일의 OHLC는 [300, 310, 300, 310] 이다. 따라서 이 거래일의 변동성은 10 (310 - 300)이다.
> 10  

In [64]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
_ = [print(x[1]-x[2]) for x in ohlc[1:] if x[-1]>x[0]]

10


### 200. 다중배열
리스트에는 3일 간의 ohlc 데이터가 저장돼 있다. 시가에 매수해서 종가에 매도 했을 경우 총 수익금을 계산하라.
```python
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
```
1일차 수익 0원 (100 - 100), 2일차 수익 -10원 (190 - 200), 3일차 수익 10원 (310 - 300) 이다.
> 0  

In [67]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]
print(sum([x[-1]-x[0] for x in ohlc[1:]]))

0


---
## 201~240. 파이썬 함수

### 201. 파이썬 함수
"비트코인" 문자열을 화면에 출력하는 print_coin() 함수를 정의하라.

In [24]:
def print_coin(): print('비트코인')
print(print_coin)

<function print_coin at 0x0000019FD14771A0>


### 202. 파이썬 함수
201번에서 정의한 함수를 호출하라.

In [25]:
print_coin()

비트코인


### 203. 파이썬 함수
201번에서 정의한 print_coin 함수를 100번호출하라.

In [26]:
_ = [print_coin() for _ in range(100)]

비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인


### 204. 파이썬 함수
"비트코인" 문자열을 100번 화면에 출력하는 print_coins() 함수를 정의하라.

In [28]:
def print_coins(): [print_coin() for _ in range(100)]
print(print_coins)

<function print_coins at 0x0000019FD14767A0>


### 205. 파이썬 함수
아래의 에러가 발생하는 이유에 대해 설명하라.
```python
hello()
def hello():
    print("Hi")
```
실행 예  
> NameError: name 'hello' is not defined  

In [30]:
# 함수 선언 후 호출해야 함
hello()
def hello():
    print("Hi")


NameError: name 'hello' is not defined

### 206. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def message() :
    print("A")
    print("B")

message()
print("C")
message()
```

In [33]:
# A
# B
# C
# A
# B
def message() :
    print("A")
    print("B")

message()
print("C")
message()

A
B
C
A
B


### 207. 파이썬 함수
아래 코드의 실행 결과를 예측하라. (읽기 어려운 코드의 예입니다.)
```python
print("A")

def message() :
    print("B")

print("C")
message()
```

In [34]:
# A
# C
# B
print("A")

def message() :
    print("B")

print("C")
message()

A
C
B


### 208. 파이썬 함수
아래 코드의 실행 결과를 예측하라. (읽기 어려운 코드의 예입니다.)
```python
print("A")
def message1() :
    print("B")
print("C")
def message2() :
    print("D")
message1()
print("E")
message2()
```

In [35]:
# A
# C
# B
# E
# D
print("A")
def message1() :
    print("B")
print("C")
def message2() :
    print("D")
message1()
print("E")
message2()

A
C
B
E
D


### 209. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def message1():
    print("A")

def message2():
    print("B")
    message1()

message2()
```

In [36]:
# B
# A
def message1():
    print("A")

def message2():
    print("B")
    message1()

message2()

B
A


### 210. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def message1():
    print("A")

def message2():
    print("B")

def message3():
    for i in range (3) :
        message2()
        print("C")
    message1()

message3()
```

In [37]:
# B
# C
# B
# C
# B
# C
# A
def message1():
    print("A")

def message2():
    print("B")

def message3():
    for i in range (3) :
        message2()
        print("C")
    message1()

message3()

B
C
B
C
B
C
A


### 211. 파이썬 함수
함수의 호출 결과를 예측하라.
```python
def 함수(문자열) :
    print(문자열)

함수("안녕")
함수("Hi")
```

In [38]:
# 안녕
# Hi
def 함수(문자열) :
    print(문자열)

함수("안녕")
함수("Hi")

안녕
Hi


### 212. 파이썬 함수
함수의 호출 결과를 예측하라.
```python
def 함수(a, b) :
    print(a + b)

함수(3, 4)
함수(7, 8)
```

In [39]:
# 7
# 15
def 함수(a, b) :
    print(a + b)

함수(3, 4)
함수(7, 8)

7
15


### 213. 파이썬 함수
아래와 같은 에러가 발생하는 원인을 설명하라.
```
def 함수(문자열) :
    print(문자열)
함수()
```
> TypeError: 함수() missing 1 required positional argument: '문자열'  

In [41]:
# 함수의 정의와 다르게 매개변수 없이 함수를 호출
def 함수(문자열) :
    print(문자열)
함수()

TypeError: 함수() missing 1 required positional argument: '문자열'

### 214. 파이썬 함수
아래와 같은 에러가 발생하는 원인을 설명하라.
```python
def 함수(a, b) :
    print(a + b)

함수("안녕", 3)
```
> TypeError: must be str, not int  

In [42]:
# 숫자 덧셈 실행시 타입 불일치로 에러
def 함수(a, b) :
    print(a + b)

함수("안녕", 3)

TypeError: can only concatenate str (not "int") to str

### 215. 파이썬 함수
하나의 문자를 입력받아 문자열 끝에 ":D" 스마일 문자열을 이어 붙여 출력하는 print_with_smile 함수를 정의하라.

In [44]:
def print_with_smile(text): print(f"{text} :D")
print(print_with_smile)

<function print_with_smile at 0x0000019FD194AC00>


### 216. 파이썬 함수
215에서 정의한 함수를 호출하라. 파라미터는 "안녕하세요"로 입력하라.

In [45]:
print_with_smile('안녕하세요')

안녕하세요 :D


### 217. 파이썬 함수
현재 가격을 입력 받아 상한가 (30%)를 출력하는 print_upper_price 함수를 정의하라.

In [47]:
def print_upper_price(price): print(price * 1.3)
print_upper_price(1000)

1300.0


### 218. 파이썬 함수
두 개의 숫자를 입력받아 두 수의 합을 출력하는 print_sum 함수를 정의하라.

In [48]:
def print_sum(num1, num2): print(num1 + num2)
print_sum(10, 20)

30


### 219. 파이썬 함수
두 개의 숫자를 입력받아 합/차/곱/나눗셈을 출력하는 print_arithmetic_operation 함수를 작성하라.
```python
print_arithmetic_operation(3, 4)
```
> 3 + 4 = 7  
> 3 - 4 = -1  
> 3 * 4 = 12  
> 3 / 4 = 0.75  

In [49]:
def print_arithmetic_operation(num1, num2):
    print(f"{num1} + {num2}  = {num1 + num2}")
    print(f"{num1} - {num2}  = {num1 - num2}")
    print(f"{num1} * {num2}  = {num1 * num2}")
    print(f"{num1} / {num2}  = {num1 / num2}")
print_arithmetic_operation(3, 4)

3 + 4  = 7
3 - 4  = -1
3 * 4  = 12
3 / 4  = 0.75


### 220. 파이썬 함수
세 개의 숫자를 입력받아 가장 큰수를 출력하는 print_max 함수를 정의하라. 단 if 문을 사용해서 수를 비교하라.

In [60]:
def print_max(num1, num2, num3): print(tmp if (tmp := num1 if num1 > num2 else num2) > num3 else num3)
print_max(1,2,3)

3


### 221. 파이썬 함수
입력된 문자열을 역순으로 출력하는 print_reverse 함수를 정의하라.
```python
print_reverse("python")
```
> nohtyp

In [61]:
def print_reverse(text): print(text[::-1])
print_reverse('python')

nohtyp


### 222. 파이썬 함수
성적 리스트를 입력 받아 평균을 출력하는 print_score 함수를 정의하라.
```python
print_score ([1, 2, 3])
```
> 2.0

In [64]:
def print_score(scoreArr): print(sum(scoreArr)/len(scoreArr))
print_score([1, 2, 3])

2.0


### 223. 파이썬 함수
하나의 리스트를 입력받아 짝수만 화면에 출력하는 print_even 함수를 정의하라.
```python
print_even ([1, 3, 2, 10, 12, 11, 15])
```
> 2  
> 10  
> 12  

In [66]:
def print_even(numArr): _ = [print(x) for x in numArr if not(x%2)]
print_even([1, 3, 2, 10, 12, 11, 15])

2
10
12


### 224. 파이썬 함수
하나의 딕셔너리를 입력받아 딕셔너리의 key 값을 화면에 출력하는 print_keys 함수를 정의하라.
```python
print_keys ({"이름":"김말똥", "나이":30, "성별":0})
```
> 이름  
> 나이  
> 성별  

In [68]:
def print_keys(dic): _ = [print(x) for x in dic.keys()]
print_keys({"이름":"김말똥", "나이":30, "성별":0})

이름
나이
성별


### 225. 파이썬 함수
my_dict에는 날짜를 키값으로 OHLC가 리스트로 저장돼 있다.
```python
my_dict = {"10/26" : [100, 130, 100, 100],
           "10/27" : [10, 12, 10, 11]}
```
my_dict와 날짜 키값을 입력받아 OHLC 리스트를 출력하는 print_value_by_key 함수를 정의하라.
```python
print_value_by_key  (my_dict, "10/26")
```
> [100, 130, 100, 100]  

In [71]:
def print_value_by_key(dic, key): print(dic[key])
my_dict = {"10/26" : [100, 130, 100, 100],
           "10/27" : [10, 12, 10, 11]}
print_value_by_key(my_dict, "10/26")

[100, 130, 100, 100]


### 226. 파이썬 함수
입력 문자열을 한 줄에 다섯글자씩 출력하는 print_5xn(string) 함수를 작성하라.
```python
print_5xn("아이엠어보이유알어걸")
```
> 아이엠어보  
> 이유알어걸  

In [78]:
def print_5xn(text): _ = [print(text[i:i+5]) for i in range(0,len(text),5)]
print_5xn("아이엠어보이유알어걸")

아이엠어보
이유알어걸


### 227. 파이썬 함수
문자열과 한줄에 출력될 글자 수를 입력을 받아 한 줄에 입력된 글자 수만큼 출력하는 print_mxn(string, number) 함수를 작성하라.
```python
print_mxn("아이엠어보이유알어걸", 3)
```
> 아이엠  
> 어보이  
> 유알어  
> 걸  

In [82]:
def print_mxn(text, n): _ = [print(text[i:i+n]) for i in range(0,len(text),n)]
print_mxn("아이엠어보이유알어걸", 3)

아이엠
어보이
유알어
걸


### 228. 파이썬 함수
연봉을 입력받아 월급을 계산하는 calc_monthly_salary(annual_salary) 함수를 정의하라. 회사는 연봉을 12개월로 나누어 분할 지급하며, 이 때 1원 미만은 버림한다.
```python
calc_monthly_salary(12000000)
```
> 1000000  

In [85]:
def calc_monthly_salary(annualSalary): print(f"{annualSalary//12}")
calc_monthly_salary(12000000)

1000000


### 229. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(a=100, b=200)
```

In [86]:
# 왼쪽: 100
# 오른쪽: 200
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(a=100, b=200)

왼쪽: 100
오른쪽: 200


### 230. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(b=100, a=200)
```

In [87]:
# 왼쪽: 200
# 오른쪽: 100
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(b=100, a=200)

왼쪽: 200
오른쪽: 100


### 231. 파이썬 함수
아래 코드를 실행한 결과를 예상하라.
```python
def n_plus_1 (n) :
    result = n + 1

n_plus_1(3)
print (result)
```

In [91]:
# 외부에서 함수 내부의 result에 접근이 불가하기에 result는 정의되지 않은 변수
def n_plus_1 (n) :
    result = n + 1

n_plus_1(3)
print (result)

NameError: name 'result' is not defined

### 232. 파이썬 함수
문자열 하나를 입력받아 인터넷 주소를 반환하는 make_url 함수를 정의하라.
```python
make_url("naver")
```
> www.naver.com  

In [92]:
make_url = lambda text: f"www.{text}.com"
print(make_url('naver'))

www.naver.com


### 233. 파이썬 함수
문자열을 입력받아 각 문자들로 구성된 리스트로 반환하는 make_list 함수를 정의하라.
```python
make_list("abcd")
```
> ['a', 'b', 'c', 'd']  

In [93]:
make_list = lambda text: list(text)
print(make_list('abcd'))

['a', 'b', 'c', 'd']


### 234. 파이썬 함수
숫자로 구성된 하나의 리스트를 입력받아, 짝수들을 추출하여 리스트로 반환하는 pickup_even 함수를 구현하라.
```python
pickup_even([3, 4, 5, 6, 7, 8])
```
> [4, 6, 8]  

In [95]:
pickup_even = lambda numArr: [x for x in numArr if not(x%2)]
print(pickup_even([3, 4, 5, 6, 7, 8]))

[4, 6, 8]


### 235. 파이썬 함수
콤마가 포함된 문자열 숫자를 입력받아 정수로 변환하는 convert_int 함수를 정의하라.
```python
convert_int("1,234,567")
```
> 1234567  

In [97]:
convert_int = lambda numStr: int(numStr.replace(',',''))
convert_int("1,234,567")

1234567

### 236. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def 함수(num) :
    return num + 4

a = 함수(10)
b = 함수(a)
c = 함수(b)
print(c)
```

In [98]:
# 22
def 함수(num) :
    return num + 4

a = 함수(10)
b = 함수(a)
c = 함수(b)
print(c)

22


### 237. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def 함수(num) :
    return num + 4

c = 함수(함수(함수(10)))
print(c)
```

In [99]:
# 22
def 함수(num) :
    return num + 4

c = 함수(함수(함수(10)))
print(c)

22


### 238. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def 함수1(num) :
    return num + 4

def 함수2(num) :
    return num * 10

a = 함수1(10)
c = 함수2(a)
print(c)
```

In [100]:
# 140
def 함수1(num) :
    return num + 4

def 함수2(num) :
    return num * 10

a = 함수1(10)
c = 함수2(a)
print(c)

140


### 239. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def 함수1(num) :
    return num + 4

def 함수2(num) :
    num = num + 2
    return 함수1(num)

c = 함수2(10)
print(c)
```

In [101]:
# 16
def 함수1(num) :
    return num + 4

def 함수2(num) :
    num = num + 2
    return 함수1(num)

c = 함수2(10)
print(c)

16


### 240. 파이썬 함수
아래 코드의 실행 결과를 예측하라.
```python
def 함수0(num) :
    return num * 2

def 함수1(num) :
    return 함수0(num + 2)

def 함수2(num) :
    num = num + 10
    return 함수1(num)

c = 함수2(2)
print(c)
```

In [102]:
# 28
def 함수0(num) :
    return num * 2

def 함수1(num) :
    return 함수0(num + 2)

def 함수2(num) :
    num = num + 10
    return 함수1(num)

c = 함수2(2)
print(c)

28


---
## 파이썬 모듈 

### 241. 현재시간
datetime 모듈을 사용해서 현재 시간을 화면에 출력해보세요.

In [15]:
from datetime import datetime as dt
print(dt.now())

2024-09-24 03:06:40.827990


### 242. 현재시간의 타입
datetime 모듈의 now 함수의 리턴 값의 타입을 화면에 출력해보세요.

In [16]:
from datetime import datetime as dt
print(type(dt.now()))

<class 'datetime.datetime'>


### 243. timedelta
datetime 모듈의 timedelta를 사용해서 오늘로부터 5일, 4일, 3일, 2일, 1일 전의 날짜를 화면에 출력해보세요.

In [18]:
import datetime as dt
_ = [print(dt.datetime.now() - dt.timedelta(n)) for n in range(5,0,-1)]

2024-09-19 03:10:16.924984
2024-09-20 03:10:16.924984
2024-09-21 03:10:16.924984
2024-09-22 03:10:16.924984
2024-09-23 03:10:16.924984


### 244. strftime
현재시간을 얻어온 후 다음과 같은 포맷으로 시간을 출력해보세요. strftime 메서드를 사용하세요.
> 18:35:01  

In [30]:
import datetime as dt
print(dt.datetime.now().strftime('%H:%M:%S'))

03:19:33


### 245. strptime
datetime.datetime.strptime 메서드를 사용하면 문자열 형식의 시간을 datetime.datetime 타입의 시간 값으로 만들어줍니다. "2020-05-04"의 문자열을 시간 타입으로 변환해보세요.

In [41]:
import datetime as dt
print(dt.datetime.strptime('2020-05-04', '%Y-%m-%d'))

2020-05-04 00:00:00


### 246. sleep 함수
time 모듈, datetime 모듈을 사용해서 1초에 한 번 현재 시간을 출력하는 코드를 작성하세요.

In [48]:
import datetime as dt
import time
for i in range(1):
    print(dt.datetime.now())
    time.sleep(1)

2024-09-24 03:30:52.012771


### 247. 모듈 임포트
모듈을 임포트하는 4가지 방식에 대해 설명해보세요.

In [ ]:
# import 모듈 [as 단축명]
# from 패키지 import 모듈 [as 단축명]
# from [패키지.]모듈 import 함수 [, 함수]
# from [패키지.]모듈 import *
# exec('import 모듈')
# 모듈명 = __import__('모듈')

### 248. os 모듈
os 모듈의 getcwd 함수를 호출하여 현재 디렉터리의 경로를 화면에 출력해보세요.

In [51]:
import os
print(os.getcwd())

C:\Users\Dexter\Source\env_py312\example\python\py300


### 249. rename 함수
바탕화면에 텍스트 파일을 하나 생성한 후 os 모듈의 rename 함수를 호출하여 해당 파일의 이름을 변경해보세요.

In [101]:
import os
# 파일 생성
with open('test.txt', 'w') as f:
    f.write('test')
# 파일이름 변경
os.rename('test.txt', 'example.txt')
# 확인
print(os.listdir())
# 파일 삭제
os.remove('example.txt')

['.ipynb_checkpoints', '01_basic_ex300.ipynb', 'example.txt']


### 250. numpy
numpy 모듈의 arange 함수를 사용해서 0.0 부터 5.0까지 0.1씩 증가하는 값을 화면에 출력해보세요.

In [103]:
import numpy as np
print(np.arange(0,5,0.1))

[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3 3.4 3.5
 3.6 3.7 3.8 3.9 4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]


---
## 파이썬 클래스 

### 251. 클래스, 객체, 인스턴스
클래스와 객체에 대해 설명해봅시다.

In [ ]:
# 클래스는 일종의 설계도로, 하나의 타입을 정의하는 방법입니다. 
# 클래스에는 관련있는 데이터와 함수를 한 데 모아 정의할 수 있습니다. 
# 클래스로 만들어진 결과물을 객체라고합니다.

### 252. 클래스 정의
비어있는 사람 (Human) 클래스를 "정의" 해보세요.

In [3]:
class Human:
    pass
print(Human())

### 253. 인스턴스 생성
사람 (Human) 클래스의 인스턴스를 "생성" 하고 이를 areum 변수로 바인딩해보세요.

In [5]:
class Human:
    pass
areum = Human()
print(areum)

### 254. 클래스 생성자-1
사람 (Human) 클래스에 "응애응애"를 출력하는 생성자를 추가하세요.
```python
areum = Human()
```
> 응애응애  

In [6]:
class Human:
    def __init__(self):
        print('응애응애')
areum = Human()

응애응애


### 255. 클래스 생성자-2
사람 (Human) 클래스에 (이름, 나이, 성별)을 받는 생성자를 추가하세요.
```python
areum = Human("아름", 25, "여자")
```

In [13]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
areum = Human('아름', 25, '여자')
print(areum.name)

아름


### 256. 인스턴스 속성에 접근
255에서 생성한 인스턴스의 이름, 나이, 성별을 출력하세요. 인스턴스 변수에 접근하여 값을 출력하면 됩니다.

> 이름: 조아름, 나이: 25, 성별: 여자

인스턴스 변수에 접근하여 값을 가져오는 예  
```python
areum.age
```
> 25  

In [14]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
areum = Human('아름', 25, '여자')
print(f'이름: {areum.name}, 나이: {areum.age}, 성별: {areum.sex}')

이름: 아름, 나이: 25, 성별: 여자


### 257. 클래스 메소드 - 1
사람 (Human) 클래스에서 이름, 나이, 성별을 출력하는 who() 메소드를 추가하세요.
```python
areum.who()
```
> 이름: 조아름, 나이: 25, 성별: 여자  

In [18]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
        
areum = Human('조아름', 25, '여자')
areum.who()

이름: 조아름, 나이: 25, 성별: 여자


### 258. 클래스 메소드 - 2
사람 (Human) 클래스에 (이름, 나이, 성별)을 받는 setInfo 메소드를 추가하세요.
```python
areum = Human("모름", 0, "모름")
areum.setInfo("아름", 25, "여자")
```

In [20]:
class Human:
    def __init__(self, name, age, sex):
        self.setInfo(name, age, sex)
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
    def setInfo(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
        
areum = Human("모름", 0, "모름")
areum.who()
areum.setInfo("아름", 25, "여자")
areum.who()

이름: 모름, 나이: 0, 성별: 모름
이름: 아름, 나이: 25, 성별: 여자


### 259. 클래스 소멸자
사람 (human) 클래스에 "나의 죽음을 알리지 말라"를 출력하는 소멸자를 추가하세요.
```python
areum = Human("아름", 25, "여자")
del areum
```
> 나의 죽음을 알리지 말라  

In [22]:
class Human:
    def __init__(self, name, age, sex):
        self.setInfo(name, age, sex)
    def __del__(self):
        print('나의 죽음을 알리지 말라')
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
    def setInfo(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex

areum = Human("아름", 25, "여자")
del(areum)

나의 죽음을 알리지 말라


### 260 에러의 원인
아래와 같은 에러가 발생한 원인에 대해 설명하세요.
```python
class OMG : 
    def print() :
        print("Oh my god")

myStock = OMG()
myStock.print()
```

> `TypeError Traceback (most recent call last)`  
> `<ipython-input-233-c85c04535b22> in <module>()`  
> `----> myStock.print()`  
> `TypeError: print() takes 0 positional arguments but 1 was given`  


In [25]:
# print() 메소드는 self인자가 없기에, 정적메소드 처럼 직접 호출해야 함 
class OMG: 
    def print():
        print("Oh my god")

myStock = OMG()
# myStock.print()
OMG.print()

Oh my god


### 261. Stock 클래스 생성
주식 종목에 대한 정보를 저장하는 Stock 클래스를 정의해보세요. 클래스는 속성과 메서드를 갖고 있지 않습니다.

### 262. 생성자
Stock 클래스의 객체가 생성될 때 종목명과 종목코드를 입력 받을 수 있도록 생성자를 정의해보세요.
```python
삼성 = Stock("삼성전자", "005930")
```

### 263. 메서드
객체에 종목명을 입력할 수 있는 set_name 메서드를 추가해보세요.
```python
a = Stock(None, None)
a.set_name("삼성전자")
```

### 264. 메서드
객체에 종목코드를 입력할 수 있는 set_code 메서드를 추가해보세요.
```python
a = Stock(None, None)
a.set_code("005930")
```

### 265. 메서드
종목명과 종목코드를 리턴하는 get_name, get_code 메서드를 추가하세요. 해당 메서드를 사용하여 종목명과 종목코드를 얻고 이를 출력해보세요.
```python
삼성 = Stock("삼성전자", "005930")
```

### 266. 객체의 속성값 업데이트
생성자에서 종목명, 종목코드, PER, PBR, 배당수익률을 입력 받을 수 있도록 생성자를 수정하세요. PER, PBR, 배당수익률은 float 타입입니다.

### 267. 객체 생성
266번에서 정의한 생성자를 통해 다음 정보를 갖는 객체를 생성해보세요.

항목|정보
--|--
종목명|삼성전자
종목코드|005930
PER|15.79
PBR|1.33
배당수익률|2.83


### 268. 객체의 속성 수정
PER, PBR, 배당수익률은 변경될 수 있는 값입니다. 이 값을 변경할 때 사용하는 set_per, set_pbr, set_dividend 메서드를 추가하세요.

### 269. 객체의 속성 수정
267번에서 생성한 객체에 set_per 메서드를 호출하여 per 값을 12.75로 수정해보세요.

### 270. 여러 종목의 객체 생성
아래의 표를 참조하여 3종목에 대해 객체를 생성하고 이를 파이썬 리스트에 저장하세요. 파이썬 리스트에 저장된 각 종목에 대해 for 루프를 통해 종목코드와 PER을 출력해보세요.

종목명|종목코드|PER|PBR|배당수익률
--|--|--|--|--
삼성전자|005930|15.79|1.33|2.83
현대차|005380|8.70|0.35|4.27
LG전자|066570|317.34|0.69|1.37

### 

### 

### 

### 

### 

### 

### 

### 

### 

### 

### 